<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jesse Parker">

<title>hybridization_MS_script</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="hybridization_MS_script_files/libs/clipboard/clipboard.min.js"></script>
<script src="hybridization_MS_script_files/libs/quarto-html/quarto.js"></script>
<script src="hybridization_MS_script_files/libs/quarto-html/popper.min.js"></script>
<script src="hybridization_MS_script_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="hybridization_MS_script_files/libs/quarto-html/anchor.min.js"></script>
<link href="hybridization_MS_script_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="hybridization_MS_script_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="hybridization_MS_script_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="hybridization_MS_script_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="hybridization_MS_script_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#set-up-copy-github-quercus_bicolor-repository-to-home-directory" id="toc-set-up-copy-github-quercus_bicolor-repository-to-home-directory" class="nav-link active" data-scroll-target="#set-up-copy-github-quercus_bicolor-repository-to-home-directory">Set-up: copy GitHub “Quercus_bicolor” repository to home directory</a></li>
  <li><a href="#i.-processing-raw-radseq-datacommand-line" id="toc-i.-processing-raw-radseq-datacommand-line" class="nav-link" data-scroll-target="#i.-processing-raw-radseq-datacommand-line">I. Processing raw RADseq data…command line</a>
  <ul class="collapse">
  <li><a href="#samples-available-on-the-sra-are-already-demultiplexed" id="toc-samples-available-on-the-sra-are-already-demultiplexed" class="nav-link" data-scroll-target="#samples-available-on-the-sra-are-already-demultiplexed">(samples available on the SRA are already demultiplexed)</a></li>
  </ul></li>
  <li><a href="#ii.-assessing-hybridization-in-r" id="toc-ii.-assessing-hybridization-in-r" class="nav-link" data-scroll-target="#ii.-assessing-hybridization-in-r">II. Assessing Hybridization (in R)</a>
  <ul class="collapse">
  <li><a href="#composed-of-three-main-parts-pca-using-dartr-package-snmf-using-lea-package-and-dsuite" id="toc-composed-of-three-main-parts-pca-using-dartr-package-snmf-using-lea-package-and-dsuite" class="nav-link" data-scroll-target="#composed-of-three-main-parts-pca-using-dartr-package-snmf-using-lea-package-and-dsuite">Composed of three main parts: PCA (using dartR package), sNMF (using LEA package), and Dsuite</a></li>
  <li><a href="#perform-principal-components-analysis-and-visualize" id="toc-perform-principal-components-analysis-and-visualize" class="nav-link" data-scroll-target="#perform-principal-components-analysis-and-visualize">Perform principal components analysis and visualize</a></li>
  <li><a href="#run-snmf-in-lea" id="toc-run-snmf-in-lea" class="nav-link" data-scroll-target="#run-snmf-in-lea">Run sNMF in LEA</a></li>
  <li><a href="#run-dsuite-command-line-to-get-pattersons-d-statistics" id="toc-run-dsuite-command-line-to-get-pattersons-d-statistics" class="nav-link" data-scroll-target="#run-dsuite-command-line-to-get-pattersons-d-statistics">Run Dsuite (command line) to get Patterson’s D statistics</a></li>
  <li><a href="#analyze-dataset-2-using-snmf" id="toc-analyze-dataset-2-using-snmf" class="nav-link" data-scroll-target="#analyze-dataset-2-using-snmf">Analyze dataset 2 using sNMF</a></li>
  <li><a href="#run-newhybrids" id="toc-run-newhybrids" class="nav-link" data-scroll-target="#run-newhybrids">Run NewHybrids</a></li>
  </ul></li>
  <li><a href="#iii.-assess-heterozygosity-fst-fis-relative-to-q.-bicolor-ancestry-proportion-v4" id="toc-iii.-assess-heterozygosity-fst-fis-relative-to-q.-bicolor-ancestry-proportion-v4" class="nav-link" data-scroll-target="#iii.-assess-heterozygosity-fst-fis-relative-to-q.-bicolor-ancestry-proportion-v4">III. Assess heterozygosity, Fst, Fis relative to Q. bicolor ancestry proportion (V4)</a></li>
  <li><a href="#iv.-assess-gd-ibd-amova-and-structure-only-including-pure-q.-bicolor" id="toc-iv.-assess-gd-ibd-amova-and-structure-only-including-pure-q.-bicolor" class="nav-link" data-scroll-target="#iv.-assess-gd-ibd-amova-and-structure-only-including-pure-q.-bicolor">IV. Assess GD, IBD, AMOVA, and structure only including “pure” Q. bicolor</a></li>
  <li><a href="#geospatial-analysis" id="toc-geospatial-analysis" class="nav-link" data-scroll-target="#geospatial-analysis">Geospatial analysis</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">hybridization_MS_script</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jesse Parker </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="set-up-copy-github-quercus_bicolor-repository-to-home-directory" class="level3">
<h3 class="anchored" data-anchor-id="set-up-copy-github-quercus_bicolor-repository-to-home-directory">Set-up: copy GitHub “Quercus_bicolor” repository to home directory</h3>
</section>
<section id="i.-processing-raw-radseq-datacommand-line" class="level1">
<h1>I. Processing raw RADseq data…command line</h1>
<section id="samples-available-on-the-sra-are-already-demultiplexed" class="level2">
<h2 class="anchored" data-anchor-id="samples-available-on-the-sra-are-already-demultiplexed">(samples available on the SRA are already demultiplexed)</h2>
<section id="create-directories" class="level4">
<h4 class="anchored" data-anchor-id="create-directories">1. Create directories</h4>
<div class="sourceCode" id="cb1" data-eval="FALSE"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">setwd</span><span class="er">(</span><span class="st">"~/Quercus_bicolor/"</span><span class="kw">)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> reference</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> demultiplexed</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> align</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> gstacks_out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="run-stacks-process_radtags-on-raw-data-make-sure-you-have-stacks-installed-in-working-directory.-can-skip-this-step-as-reads-available-from-sra-are-already-demultiplexed" class="level4">
<h4 class="anchored" data-anchor-id="run-stacks-process_radtags-on-raw-data-make-sure-you-have-stacks-installed-in-working-directory.-can-skip-this-step-as-reads-available-from-sra-are-already-demultiplexed">2. Run STACKS process_radtags on raw data (make sure you have STACKS installed in working directory). can skip this step as reads available from SRA are already demultiplexed</h4>
<div class="sourceCode" id="cb2" data-eval="FALSE"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">stacks-2.68/process_radtags</span> <span class="at">-P</span> <span class="at">-p</span> ./data/run_1/ <span class="at">-o</span> ./demultiplexed/ <span class="at">-b</span> ./data/C1591A_barcodes.tsv  <span class="at">-e</span> pstI <span class="at">-r</span> <span class="at">--inline-null</span> <span class="at">-c</span> <span class="at">-q</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">stacks-2.68/process_radtags</span> <span class="at">-P</span> <span class="at">-p</span> ./data/run_2/ <span class="at">-o</span> ./demultiplexed/ <span class="at">-b</span> ./data/C1591B_barcodes.tsv  <span class="at">-e</span> pstI <span class="at">-r</span> <span class="at">--inline-null</span> <span class="at">-c</span> <span class="at">-q</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="download-using-sra-toolkitmust-be-installed-and-index-reference-genome-first-load-bwa-with-conda" class="level4">
<h4 class="anchored" data-anchor-id="download-using-sra-toolkitmust-be-installed-and-index-reference-genome-first-load-bwa-with-conda">3. Download (using SRA toolkit…must be installed) and index reference genome (first load BWA with conda)</h4>
<div class="sourceCode" id="cb3" data-eval="FALSE"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> reference</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">datasets</span> download genome accession GCA_036321655.1 <span class="at">--include</span> genome</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> index ./reference/GCA_036321655.1_ASM3632165v1_genomic.fna.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="create-script-to-iterate-through-samples-and-align-to-reference-genome" class="level4">
<h4 class="anchored" data-anchor-id="create-script-to-iterate-through-samples-and-align-to-reference-genome">4. Create script to iterate through samples and align to reference genome</h4>
<div class="sourceCode" id="cb4" data-eval="FALSE"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nano</span> align_samples.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="modify-script-with-the-following" class="level5">
<h5 class="anchored" data-anchor-id="modify-script-with-the-following">Modify script with the following:</h5>
<p><code>{.bash eval= FALSE} #!/bin/bash REF="./reference/GCA_036321655.1_ASM3632165v1_genomic.fna.gz" #Create an array of sample names (removes .1.fq.gz suffix) INDS=($(for i in ./demultiplexed/*.1.fq.gz; do echo $(basename ${i%.1.fq.gz}); done)) #Print the list of sample names echo "INDS array contents: ${INDS[@]}" #Loop through each sample and set the variables for IND in ${INDS[@]}; do         # Declare file paths for forward and reverse reads, and output BAM file         FORWARD=./demultiplexed/samples/${IND}.1.fq.gz         REVERSE=./demultiplexed/samples/${IND}.2.fq.gz         OUTPUT=./align/${IND}_sort.bam         echo "Aligning $IND with bwa"         bwa mem -t 4 $REF $FORWARD $REVERSE | samtools view -b | samtools sort -T ${IND} &gt; $OUTPUT done</code></p>
</section>
<section id="run-script-with" class="level5">
<h5 class="anchored" data-anchor-id="run-script-with">Run script with:</h5>
<p><code>{.bash eval= FALSE} bash align_samples.sh</code></p>
</section>
<section id="repeat-for-reference-samples-modifying-as-necessary" class="level5">
<h5 class="anchored" data-anchor-id="repeat-for-reference-samples-modifying-as-necessary">Repeat for reference samples, modifying as necessary</h5>
</section>
</section>
<section id="run-the-stacks-gstacks-module-on-all-aligned-samples" class="level4">
<h4 class="anchored" data-anchor-id="run-the-stacks-gstacks-module-on-all-aligned-samples">5. Run the STACKS gstacks module on all aligned samples</h4>
<p><code>{.bash eval= FALSE} stacks-2.68/gstacks -I ./align/ -M ./data/popmap_gstacks.txt --min-mapq 30 -O ./gstacks_out/ -t 8</code></p>
</section>
<section id="run-the-stacks-populations-module" class="level4">
<h4 class="anchored" data-anchor-id="run-the-stacks-populations-module">6. Run the STACKS populations module</h4>
<div class="sourceCode" id="cb5" data-eval="false"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> gstacks_out/dataset1</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">stacks-2.68/populations</span> <span class="at">--in-path</span> ./gstacks_out/ <span class="at">--popmap</span> ./data/popmap_dataset1.txt <span class="at">--out-path</span> ./data/STACKS/dataset1 <span class="at">-R</span> 0.95 <span class="at">--min-gt-depth</span> 5 <span class="at">--min-mac</span> 3 <span class="at">--vcf</span> <span class="at">--structure</span> <span class="at">--ordered-export</span> <span class="at">--write-single-snp</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="ex">stacks-2.68/populations</span> <span class="at">--in-path</span> ./gstacks_out/ <span class="at">--popmap</span> ./data/popmap_dataset2.txt <span class="at">--out-path</span> ./data/STACKS/dataset2 <span class="at">-R</span> 0.95 <span class="at">--min-gt-depth</span> 5 <span class="at">--min-mac</span> 3 <span class="at">--vcf</span> <span class="at">--structure</span> <span class="at">--ordered-export</span> <span class="at">--write-single-snp</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="ex">stacks-2.68/populations</span> <span class="at">--in-path</span> ./gstacks_out/ <span class="at">--popmap</span> ./data/popmap_dataset2.txt <span class="at">--out-path</span> ./data/STACKS/dataset2_NoMissing/ <span class="at">-R</span> 1 <span class="at">--min-gt-depth</span> 5 <span class="at">--min-mac</span> 3 <span class="at">--vcf</span> <span class="at">--ordered-export</span> <span class="at">--write-single-snp</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="you-now-have-three-.vcf-files-of-filtered-snps-that-can-be-used-for-further-analyses-in-r.-one-includes-all-the-species-dataset1-another-only-includes-q.bicolor-and-q.lyrata-dataset2-with-0.95-call-rate-and-the-third-only-includes-q.bicolor-and-q.lyrata-dataset2-with-1-call-rate" class="level5">
<h5 class="anchored" data-anchor-id="you-now-have-three-.vcf-files-of-filtered-snps-that-can-be-used-for-further-analyses-in-r.-one-includes-all-the-species-dataset1-another-only-includes-q.bicolor-and-q.lyrata-dataset2-with-0.95-call-rate-and-the-third-only-includes-q.bicolor-and-q.lyrata-dataset2-with-1-call-rate">You now have three .vcf files of filtered SNPs that can be used for further analyses in R. One includes all the species (dataset1), another only includes Q.bicolor and Q.lyrata (dataset2) with 0.95 call rate, and the third only includes Q.bicolor and Q.lyrata (dataset2) with 1 call rate</h5>
</section>
</section>
</section>
</section>
<section id="ii.-assessing-hybridization-in-r" class="level1">
<h1>II. Assessing Hybridization (in R)</h1>
<section id="composed-of-three-main-parts-pca-using-dartr-package-snmf-using-lea-package-and-dsuite" class="level2">
<h2 class="anchored" data-anchor-id="composed-of-three-main-parts-pca-using-dartr-package-snmf-using-lea-package-and-dsuite">Composed of three main parts: PCA (using dartR package), sNMF (using LEA package), and Dsuite</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="load-.vcf-file-from-stacks-populations-and-add-population-names" class="level4">
<h4 class="anchored" data-anchor-id="load-.vcf-file-from-stacks-populations-and-add-population-names">Load .vcf file from STACKS populations and add population names</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(adegenet)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LEA)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vcfR)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(poppr)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>vc<span class="ot">&lt;-</span><span class="fu">read.vcfR</span>(<span class="st">"./data/STACKS/dataset1/populations.snps.vcf"</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">vcfR2genlight</span>(vc)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">gl.compliance.check</span>(g)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>names<span class="ot">&lt;-</span>g<span class="sc">@</span>ind.names</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>pops<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">"Qbicolor_161_pops.csv"</span>, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>pop<span class="ot">&lt;-</span>pops<span class="sc">$</span>pop <span class="sc">%&gt;%</span> <span class="fu">as.factor</span>()</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>pop<span class="ot">&lt;-</span>pop</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="perform-principal-components-analysis-and-visualize" class="level2">
<h2 class="anchored" data-anchor-id="perform-principal-components-analysis-and-visualize">Perform principal components analysis and visualize</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot) </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">glPca</span>(g, <span class="at">center=</span><span class="cn">TRUE</span>, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co">#for all samples, keep 7</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(pca<span class="sc">$</span>eig, <span class="at">main=</span><span class="st">"eigenvalues"</span>, <span class="at">col=</span><span class="fu">heat.colors</span>(<span class="fu">length</span>(pca<span class="sc">$</span>eig)))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a dataframe for plotting</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>pca_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC1 =</span> pca<span class="sc">$</span>scores[,<span class="dv">1</span>],</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC2 =</span> pca<span class="sc">$</span>scores[,<span class="dv">2</span>],</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">Population =</span> <span class="fu">as.factor</span>(<span class="fu">pop</span>(g))  <span class="co"># Population as categorical variable</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(pca_df[,<span class="dv">3</span>]) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">levels</span>(pca_df[,<span class="dv">3</span>]), <span class="st">"Sample (outlier)"</span>)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Now you can assign it</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>pca_df[<span class="dv">1</span>,<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="st">"Sample (outlier)"</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the color palette</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">'viridis'</span>)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>color_palette <span class="ot">&lt;-</span> <span class="fu">viridis</span>(<span class="dv">29</span>, <span class="at">option=</span><span class="st">'turbo'</span>) </span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Select one representative per population for labeling</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>pca_labels <span class="ot">&lt;-</span> pca_df <span class="sc">%&gt;%</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Population) <span class="sc">%&gt;%</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="dv">1</span>)  <span class="co"># Take the first occurrence per population</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge to get label positions for each point</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>pca_df <span class="ot">&lt;-</span> pca_df <span class="sc">%&gt;%</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(pca_labels, <span class="at">by =</span> <span class="st">"Population"</span>, <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">""</span>, <span class="st">"_label"</span>))</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Variance explained (percentages for axis labels)</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>var_exp1 <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> pca<span class="sc">$</span>eig[<span class="dv">1</span>] <span class="sc">/</span> <span class="fu">sum</span>(pca<span class="sc">$</span>eig), <span class="dv">2</span>)</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>var_exp2 <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> pca<span class="sc">$</span>eig[<span class="dv">2</span>] <span class="sc">/</span> <span class="fu">sum</span>(pca<span class="sc">$</span>eig), <span class="dv">2</span>)</span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>pca_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_df, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">color =</span> Population)) <span class="sc">+</span></span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>  <span class="co">#geom_segment(aes(x = PC1_label, y = PC2_label, xend = PC1, yend = PC2), </span></span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             color = "gray50", alpha = 0.5, linetype = "dotted") +  </span></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">4</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span>  </span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> pca_labels,</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> Population),</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="dv">5</span>,</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">family =</span> <span class="st">"mono"</span>,</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">max.overlaps =</span> <span class="cn">Inf</span>,</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">nudge_x =</span> <span class="fl">0.01</span> <span class="sc">*</span> (<span class="fu">max</span>(pca_df<span class="sc">$</span>PC1) <span class="sc">-</span> <span class="fu">min</span>(pca_df<span class="sc">$</span>PC1)),  </span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>    <span class="at">nudge_y =</span> <span class="fl">0.01</span> <span class="sc">*</span> (<span class="fu">max</span>(pca_df<span class="sc">$</span>PC2) <span class="sc">-</span> <span class="fu">min</span>(pca_df<span class="sc">$</span>PC2)),  </span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">force =</span> <span class="dv">2</span>,  </span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">segment.size =</span> <span class="fl">0.7</span>,  </span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>    <span class="at">box.padding =</span> <span class="fl">1.2</span>,</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> color_palette) <span class="sc">+</span>  </span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span>  </span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="fl">0.1</span>),  <span class="co"># Lighter major grid lines</span></span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="fl">0.1</span>))<span class="sc">+</span>  </span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">size =</span> <span class="dv">18</span>, <span class="at">color=</span><span class="st">'black'</span>),  </span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">"right"</span></span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"PCA of SNP data, Dataset 3"</span>,</span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"PC1 ("</span>, var_exp1, <span class="st">"% variance)"</span>),</span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"PC2 ("</span>, var_exp2, <span class="st">"% variance)"</span>)</span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>())  </span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-66"><a href="#cb8-66" aria-hidden="true" tabindex="-1"></a><span class="co"># Scree Plot (Eigenvalues Barplot)</span></span>
<span id="cb8-67"><a href="#cb8-67" aria-hidden="true" tabindex="-1"></a>scree_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PC =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pca<span class="sc">$</span>eig), <span class="at">Eigenvalue =</span> pca<span class="sc">$</span>eig)</span>
<span id="cb8-68"><a href="#cb8-68" aria-hidden="true" tabindex="-1"></a>scree_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(scree_df, <span class="fu">aes</span>(<span class="at">x =</span> PC, <span class="at">y =</span> Eigenvalue)) <span class="sc">+</span></span>
<span id="cb8-69"><a href="#cb8-69" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb8-70"><a href="#cb8-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb8-71"><a href="#cb8-71" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="dv">0</span>),  <span class="co"># Lighter major grid lines</span></span>
<span id="cb8-72"><a href="#cb8-72" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="dv">0</span>))<span class="sc">+</span>  </span>
<span id="cb8-73"><a href="#cb8-73" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">50</span>)<span class="sc">+</span></span>
<span id="cb8-74"><a href="#cb8-74" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Scree Plot"</span>, <span class="at">x =</span> <span class="st">"Principal Component"</span>, <span class="at">y =</span> <span class="st">"Eigenvalue"</span>) <span class="sc">+</span></span>
<span id="cb8-75"><a href="#cb8-75" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb8-76"><a href="#cb8-76" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">family =</span> <span class="st">"mono"</span>),</span>
<span id="cb8-77"><a href="#cb8-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">size =</span> <span class="dv">10</span>, <span class="at">color=</span><span class="st">"black"</span>)</span>
<span id="cb8-78"><a href="#cb8-78" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb8-79"><a href="#cb8-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-80"><a href="#cb8-80" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine PCA plot with Scree Plot in bottom corner</span></span>
<span id="cb8-81"><a href="#cb8-81" aria-hidden="true" tabindex="-1"></a>final_plot <span class="ot">&lt;-</span> <span class="fu">ggdraw</span>() <span class="sc">+</span></span>
<span id="cb8-82"><a href="#cb8-82" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_plot</span>(pca_plot) <span class="sc">+</span></span>
<span id="cb8-83"><a href="#cb8-83" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_plot</span>(scree_plot, <span class="at">x =</span> .<span class="dv">1</span>, <span class="at">y =</span> <span class="fl">0.1</span>, <span class="at">width =</span> <span class="fl">0.17</span>, <span class="at">height =</span> <span class="fl">0.25</span>)</span>
<span id="cb8-84"><a href="#cb8-84" aria-hidden="true" tabindex="-1"></a>final_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-snmf-in-lea" class="level2">
<h2 class="anchored" data-anchor-id="run-snmf-in-lea">Run sNMF in LEA</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/data/"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gl2geno</span>(g, <span class="at">outfile=</span> <span class="st">"g_geno"</span>,<span class="at">outpath=</span> <span class="fu">getwd</span>())</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">#detect admixture</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>project <span class="ot">=</span> <span class="fu">snmf</span>(<span class="st">"g_geno.geno"</span>, <span class="at">K =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>, <span class="at">entropy =</span> <span class="cn">TRUE</span>, <span class="at">repetitions =</span> <span class="dv">100</span>, <span class="at">seed=</span><span class="dv">42</span>, <span class="at">project =</span> <span class="st">"new"</span>, <span class="at">iterations=</span><span class="dv">1000000</span>, <span class="at">alpha=</span><span class="dv">10</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>project <span class="ot">=</span> <span class="fu">load.snmfProject</span>(<span class="st">"g_geno.snmfProject"</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(project, <span class="at">col =</span> <span class="st">"blue"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">1.2</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co"># select the best run for K = 4 clusters </span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>best <span class="ot">=</span> <span class="fu">which.min</span>(<span class="fu">cross.entropy</span>(project, <span class="at">K =</span> <span class="dv">5</span>)) </span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"tomato"</span>, <span class="st">"olivedrab"</span>, <span class="st">"gold"</span>,<span class="st">"lightblue"</span>,<span class="st">"pink"</span>,<span class="st">"purple"</span>,<span class="st">"orange"</span>) </span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="fu">windows</span>()</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="fu">barchart</span>(project, <span class="at">K =</span> <span class="dv">5</span>, <span class="at">run =</span> best, <span class="at">border =</span> <span class="cn">NA</span>,<span class="at">sort.by.Q=</span><span class="cn">FALSE</span>, <span class="at">space =</span> <span class="dv">0</span>, <span class="at">col =</span> my.colors, <span class="at">xlab =</span> <span class="st">"Individuals"</span>, <span class="at">ylab =</span> <span class="st">"Ancestry proportions"</span>, <span class="at">main =</span> <span class="st">"Ancestry matrix"</span>)<span class="ot">-&gt;</span> bp </span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(names), <span class="at">labels =</span> names, <span class="at">las=</span><span class="dv">2</span>, <span class="at">cex.axis =</span> .<span class="dv">5</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>q_scores<span class="ot">&lt;-</span><span class="fu">Q</span>(project, <span class="at">K =</span> <span class="dv">5</span>, <span class="fu">which.min</span>(<span class="fu">cross.entropy</span>(project, <span class="at">K =</span> <span class="dv">5</span>)))</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>q<span class="ot">&lt;-</span><span class="fu">cbind</span>(pops,q_scores) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>q<span class="sc">$</span>V4<span class="ot">&lt;-</span><span class="fu">as.numeric</span>(q<span class="sc">$</span>V4)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="co">#apply threshold</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>q<span class="sc">$</span>pure<span class="ot">&lt;-</span>q<span class="sc">$</span>V4<span class="sc">&gt;</span><span class="fl">0.84</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(q,<span class="st">"qscores_k5.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="make-txt-file-for-stacks-populations-with-pure-individuals-for-analysis-in-part-4" class="level4">
<h4 class="anchored" data-anchor-id="make-txt-file-for-stacks-populations-with-pure-individuals-for-analysis-in-part-4">Make txt file for STACKS populations with pure individuals for analysis in Part 4</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>pure<span class="ot">&lt;-</span>q[q<span class="sc">$</span>pure<span class="sc">==</span><span class="cn">TRUE</span>,]</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#remove reference samples</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>pure<span class="ot">&lt;-</span>pure[pure<span class="sc">$</span>pop<span class="sc">!=</span><span class="st">'Ref. Sample-Q. bicolor'</span>,]</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/data/"</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(pure[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="st">"pure.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>, <span class="at">col.names=</span><span class="cn">FALSE</span>, <span class="at">quote=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="then-you-can-run-stacks-again-in-the-command-line-from-inside-quercus_bicolor-directory-to-generate-another-.vcf-of-the-pure-q.-bicolor-individuals-for-part-iv-below." class="level4">
<h4 class="anchored" data-anchor-id="then-you-can-run-stacks-again-in-the-command-line-from-inside-quercus_bicolor-directory-to-generate-another-.vcf-of-the-pure-q.-bicolor-individuals-for-part-iv-below.">Then you can run STACKS again in the command line (from inside “Quercus_bicolor” directory) to generate another .vcf of the “pure” Q. bicolor individuals for part IV (below).</h4>
<p><code>{.bash eval= FALSE} stacks-2.68/populations --in-path ./data/STACKS/ --popmap ./data/pure.txt --out-path ./data/STACKS/dataset3/ -R 0.95 --min-gt-depth 5 --min-mac 3 --genepop --structure --ordered-export --write-single-snp</code></p>
</section>
</section>
<section id="run-dsuite-command-line-to-get-pattersons-d-statistics" class="level2">
<h2 class="anchored" data-anchor-id="run-dsuite-command-line-to-get-pattersons-d-statistics">Run Dsuite (command line) to get Patterson’s D statistics</h2>
<section id="first-generate-new-.vcf-file-with-no-missing-data" class="level4">
<h4 class="anchored" data-anchor-id="first-generate-new-.vcf-file-with-no-missing-data">First, generate new .vcf file with no missing data</h4>
<p>``` {.bash eval= FALSE} #generate new .vcf with no missing data stacks-2.68/populations –in-path ./data/STACKS/ –popmap ./data/popmap_dataset1.txt –out-path ./data/STACKS/Dsuite/ -R 1 –min-gt-depth 5 –min-mac 3 –vcf –ordered-export –write-single-snp</p>
<p>#zip .vcf file cd ./data/STACKS/Dsuite/ gzip populations.snps.vcf cd ~/Quercus_bicolor</p>
<pre><code>#### Then, run Dsuite
``` {.bash eval= FALSE}
# after installing Dsuite in main directory
./Dsuite/Build/Dsuite Dinvestigate -n ./data/dinvestigate1 ./data/STACKS/Dsuite/populations.snps.vcf.gz ./data/SETS_2.txt ./data/test_trios_2.txt &gt; ./data/dinvestigate1_output.txt 2&gt; ./data/dinvestigate1_log.txt</code></pre>
</section>
<section id="convert-raw-dsuite-dinvestigate-output-to-dataframe" class="level4">
<h4 class="anchored" data-anchor-id="convert-raw-dsuite-dinvestigate-output-to-dataframe">Convert raw Dsuite Dinvestigate output to dataframe</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get file from Dsuite Dinvestigate</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in lines from output file</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>lines <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">"./data/dinvestigate1_output.txt"</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(lines, <span class="dv">20</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>lines<span class="ot">&lt;-</span>lines[<span class="dv">18</span><span class="sc">:</span><span class="fu">length</span>(lines)]</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize list to hold parsed results</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># We'll iterate in chunks of 7 lines (6 lines of data + 1 blank)</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> (i <span class="sc">&lt;=</span> <span class="fu">length</span>(lines) <span class="sc">-</span> <span class="dv">6</span>) {</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Skip if the current line is empty</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (lines[i] <span class="sc">==</span> <span class="st">""</span>) {</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    i <span class="ot">&lt;-</span> i <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">next</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract trio</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  trio <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(lines[i], <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>))</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>  P1 <span class="ot">&lt;-</span> trio[<span class="dv">1</span>]</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  P2 <span class="ot">&lt;-</span> trio[<span class="dv">2</span>]</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  P3 <span class="ot">&lt;-</span> trio[<span class="dv">3</span>]</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract values</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>  D       <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">"D="</span>, <span class="st">""</span>, lines[i <span class="sc">+</span> <span class="dv">1</span>]))</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>  f_d     <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">"f_d="</span>, <span class="st">""</span>, <span class="fu">strsplit</span>(lines[i <span class="sc">+</span> <span class="dv">2</span>], <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)[[<span class="dv">1</span>]][<span class="dv">1</span>]))</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>  f_dM    <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">"f_dM="</span>, <span class="st">""</span>, <span class="fu">strsplit</span>(lines[i <span class="sc">+</span> <span class="dv">3</span>], <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)[[<span class="dv">1</span>]][<span class="dv">1</span>]))</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  ABBA_p  <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">"ABBA_KSpval = "</span>, <span class="st">""</span>, lines[i <span class="sc">+</span> <span class="dv">4</span>]))</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>  BABA_p  <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">"BABA_KSpval = "</span>, <span class="st">""</span>, lines[i <span class="sc">+</span> <span class="dv">5</span>]))</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Save to list</span></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>  results[[<span class="fu">length</span>(results) <span class="sc">+</span> <span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">P1 =</span> P1, <span class="at">P2 =</span> P2, <span class="at">P3 =</span> P3,</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">D =</span> D, <span class="at">f_d =</span> f_d, <span class="at">f_dM =</span> f_dM,</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">ABBA_KSpval =</span> ABBA_p, <span class="at">BABA_KSpval =</span> BABA_p,</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Move to next block</span></span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>  i <span class="ot">&lt;-</span> i <span class="sc">+</span> <span class="dv">7</span></span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, results)</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df)[<span class="fu">colnames</span>(df) <span class="sc">==</span> <span class="st">"P2"</span>] <span class="ot">&lt;-</span> <span class="st">"names"</span></span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pops)[<span class="fu">colnames</span>(pops) <span class="sc">==</span> <span class="st">"ID"</span>] <span class="ot">&lt;-</span> <span class="st">"names"</span></span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>dfmer<span class="ot">&lt;-</span><span class="fu">merge</span>(df,q,<span class="at">by=</span><span class="st">"names"</span>)</span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>dfmer<span class="ot">&lt;-</span><span class="fu">merge</span>(dfmer, pops, <span class="at">by=</span><span class="st">'names'</span>)</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(dfmer)</span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x=</span>dfmer<span class="sc">$</span>D, <span class="at">y=</span>dfmer<span class="sc">$</span>LAT)</span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a>plot<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dfmer, <span class="fu">aes</span>(<span class="at">x =</span> D, <span class="at">y =</span> LAT)) <span class="sc">+</span></span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop.y)) <span class="sc">+</span>  <span class="co"># Color points by "pop"</span></span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"sNMF Ancestry Proportion"</span>) <span class="sc">+</span></span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="st">'D'</span>)) <span class="sc">+</span></span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"D Stat as Function of Ancestry Proportion"</span>, <span class="at">color =</span> <span class="st">"Population"</span>) <span class="sc">+</span></span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>, <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color=</span><span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),  <span class="co"># Lighter major grid lines</span></span>
<span id="cb12-57"><a href="#cb12-57" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="fl">0.05</span>))</span>
<span id="cb12-58"><a href="#cb12-58" aria-hidden="true" tabindex="-1"></a><span class="er">)</span></span>
<span id="cb12-59"><a href="#cb12-59" aria-hidden="true" tabindex="-1"></a>plot</span>
<span id="cb12-60"><a href="#cb12-60" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb12-61"><a href="#cb12-61" aria-hidden="true" tabindex="-1"></a>popD<span class="ot">&lt;-</span>dfmer <span class="sc">%&gt;%</span></span>
<span id="cb12-62"><a href="#cb12-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pop.y) <span class="sc">%&gt;%</span></span>
<span id="cb12-63"><a href="#cb12-63" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">mean_D =</span> <span class="fu">mean</span>(D, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb12-64"><a href="#cb12-64" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(popD)[<span class="fu">colnames</span>(popD) <span class="sc">==</span> <span class="st">"pop.y"</span>] <span class="ot">&lt;-</span> <span class="st">"pop"</span></span>
<span id="cb12-65"><a href="#cb12-65" aria-hidden="true" tabindex="-1"></a>merged_data<span class="ot">&lt;-</span><span class="fu">merge</span>(merged_data,popD, <span class="at">by=</span><span class="st">'pop'</span>)</span>
<span id="cb12-66"><a href="#cb12-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-67"><a href="#cb12-67" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb12-68"><a href="#cb12-68" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb12-69"><a href="#cb12-69" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate mean and SD</span></span>
<span id="cb12-70"><a href="#cb12-70" aria-hidden="true" tabindex="-1"></a>summary_stats <span class="ot">&lt;-</span> dfmer <span class="sc">%&gt;%</span></span>
<span id="cb12-71"><a href="#cb12-71" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pop.y) <span class="sc">%&gt;%</span></span>
<span id="cb12-72"><a href="#cb12-72" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(</span>
<span id="cb12-73"><a href="#cb12-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_D =</span> <span class="fu">mean</span>(D, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb12-74"><a href="#cb12-74" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_D =</span> <span class="fu">sd</span>(D, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-75"><a href="#cb12-75" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb12-76"><a href="#cb12-76" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">ifelse</span>(<span class="fu">row_number</span>() <span class="sc">&lt;=</span> <span class="dv">9</span>, <span class="st">"CORE"</span>, <span class="st">"EDGE"</span>))</span>
<span id="cb12-77"><a href="#cb12-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-78"><a href="#cb12-78" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with color by group</span></span>
<span id="cb12-79"><a href="#cb12-79" aria-hidden="true" tabindex="-1"></a>plot<span class="ot">&lt;-</span><span class="fu">ggplot</span>(summary_stats, <span class="fu">aes</span>(<span class="at">x =</span> pop.y, <span class="at">y =</span> mean_D, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb12-80"><a href="#cb12-80" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="at">width =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb12-81"><a href="#cb12-81" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> mean_D <span class="sc">-</span> sd_D, <span class="at">ymax =</span> mean_D <span class="sc">+</span> sd_D),</span>
<span id="cb12-82"><a href="#cb12-82" aria-hidden="true" tabindex="-1"></a>                <span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb12-83"><a href="#cb12-83" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"CORE"</span> <span class="ot">=</span> <span class="st">"tomato"</span>, <span class="st">"EDGE"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>)) <span class="sc">+</span></span>
<span id="cb12-84"><a href="#cb12-84" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Mean D by Population"</span>,</span>
<span id="cb12-85"><a href="#cb12-85" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Population"</span>, <span class="at">y =</span> <span class="st">"Mean D"</span>, <span class="at">fill =</span> <span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb12-86"><a href="#cb12-86" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>)) <span class="sc">+</span></span>
<span id="cb12-87"><a href="#cb12-87" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb12-88"><a href="#cb12-88" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb12-89"><a href="#cb12-89" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb12-90"><a href="#cb12-90" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">"right"</span>,</span>
<span id="cb12-91"><a href="#cb12-91" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color =</span> <span class="st">"black"</span>),</span>
<span id="cb12-92"><a href="#cb12-92" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),</span>
<span id="cb12-93"><a href="#cb12-93" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),</span>
<span id="cb12-94"><a href="#cb12-94" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)</span>
<span id="cb12-95"><a href="#cb12-95" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb12-96"><a href="#cb12-96" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span>
<span id="cb12-97"><a href="#cb12-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-98"><a href="#cb12-98" aria-hidden="true" tabindex="-1"></a><span class="co">#plot for population level</span></span>
<span id="cb12-99"><a href="#cb12-99" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>pop2<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"EDGE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>,<span class="st">"CORE"</span>)</span>
<span id="cb12-100"><a href="#cb12-100" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"tidyverse"</span>)</span>
<span id="cb12-101"><a href="#cb12-101" aria-hidden="true" tabindex="-1"></a><span class="co"># Reshape the data to long format for plotting D, f_d, and f_dM</span></span>
<span id="cb12-102"><a href="#cb12-102" aria-hidden="true" tabindex="-1"></a>df_long <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb12-103"><a href="#cb12-103" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(P2, D, f_d, f_dM, pop2) <span class="sc">%&gt;%</span></span>
<span id="cb12-104"><a href="#cb12-104" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(D, f_d, f_dM), <span class="at">names_to =</span> <span class="st">"stat"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>)</span>
<span id="cb12-105"><a href="#cb12-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-106"><a href="#cb12-106" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb12-107"><a href="#cb12-107" aria-hidden="true" tabindex="-1"></a>plot<span class="ot">&lt;-</span><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> P2, <span class="at">y =</span> D, <span class="at">fill =</span> pop2)) <span class="sc">+</span></span>
<span id="cb12-108"><a href="#cb12-108" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="at">position =</span> <span class="st">"dodge"</span>, <span class="at">width =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb12-109"><a href="#cb12-109" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"CORE"</span> <span class="ot">=</span> <span class="st">"tomato"</span>, <span class="st">"EDGE"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>)) <span class="sc">+</span></span>
<span id="cb12-110"><a href="#cb12-110" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Patterson's D Statistic by Population"</span>,</span>
<span id="cb12-111"><a href="#cb12-111" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Population"</span>, <span class="at">y =</span> <span class="st">"Value"</span>, <span class="at">fill =</span> <span class="st">"Group"</span>) <span class="sc">+</span></span>
<span id="cb12-112"><a href="#cb12-112" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb12-113"><a href="#cb12-113" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb12-114"><a href="#cb12-114" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color =</span> <span class="st">"black"</span>),</span>
<span id="cb12-115"><a href="#cb12-115" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),</span>
<span id="cb12-116"><a href="#cb12-116" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),</span>
<span id="cb12-117"><a href="#cb12-117" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb12-118"><a href="#cb12-118" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">"right"</span></span>
<span id="cb12-119"><a href="#cb12-119" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb12-120"><a href="#cb12-120" aria-hidden="true" tabindex="-1"></a>plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="analyze-dataset-2-using-snmf" class="level2">
<h2 class="anchored" data-anchor-id="analyze-dataset-2-using-snmf">Analyze dataset 2 using sNMF</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>vc<span class="ot">&lt;-</span><span class="fu">read.vcfR</span>(<span class="st">"./data/STACKS/dataset2/populations.snps.vcf"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">vcfR2genlight</span>(vc)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>pm<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">"./data/popmap_dataset2_version2.csv"</span>,<span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>pop<span class="ot">&lt;-</span>pm<span class="sc">$</span>pop <span class="sc">%&gt;%</span> <span class="fu">as.factor</span>()</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/data/"</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">gl2geno</span>(g, <span class="at">outfile=</span> <span class="st">"g_geno_dataset2"</span>,<span class="at">outpath=</span> <span class="fu">getwd</span>())</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>names<span class="ot">&lt;-</span>g<span class="sc">@</span>ind.names</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>pop<span class="ot">&lt;-</span>pm<span class="sc">$</span>pop</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co">#detect admixture</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>project <span class="ot">=</span> <span class="fu">snmf</span>(<span class="st">"g_geno_dataset2.geno"</span>, <span class="at">K =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>, <span class="at">entropy =</span> <span class="cn">TRUE</span>, <span class="at">repetitions =</span> <span class="dv">100</span>, <span class="at">seed=</span><span class="dv">42</span>, <span class="at">project =</span> <span class="st">"new"</span>, <span class="at">iterations=</span><span class="dv">1000000</span>, <span class="at">alpha=</span><span class="dv">10</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>project <span class="ot">=</span> <span class="fu">load.snmfProject</span>(<span class="st">"g_geno_dataset2.snmfProject"</span>)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(project, <span class="at">col =</span> <span class="st">"blue"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">1.2</span>)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="co"># select the best run for K = 2 clusters </span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>best <span class="ot">=</span> <span class="fu">which.min</span>(<span class="fu">cross.entropy</span>(project, <span class="at">K =</span> <span class="dv">2</span>)) </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"olivedrab"</span>, <span class="st">"lightblue"</span>) </span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a><span class="fu">windows</span>()</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="fu">barchart</span>(project, <span class="at">K =</span> <span class="dv">2</span>, <span class="at">run =</span> best, <span class="at">border =</span> <span class="cn">NA</span>,<span class="at">sort.by.Q=</span><span class="cn">FALSE</span>, <span class="at">space =</span> <span class="dv">0</span>, <span class="at">col =</span> my.colors, <span class="at">xlab =</span> <span class="st">"Individuals"</span>, <span class="at">ylab =</span> <span class="st">"Ancestry proportions"</span>, <span class="at">main =</span> <span class="st">"Ancestry matrix"</span>)<span class="ot">-&gt;</span> bp </span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(names), <span class="at">labels =</span> names, <span class="at">las=</span><span class="dv">2</span>, <span class="at">cex.axis =</span> .<span class="dv">5</span>)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>q_scores<span class="ot">&lt;-</span><span class="fu">Q</span>(project, <span class="at">K =</span> <span class="dv">2</span>, <span class="fu">which.min</span>(<span class="fu">cross.entropy</span>(project, <span class="at">K =</span> <span class="dv">5</span>)))</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>q<span class="ot">&lt;-</span><span class="fu">cbind</span>(names, pop,q_scores) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-newhybrids" class="level2">
<h2 class="anchored" data-anchor-id="run-newhybrids">Run NewHybrids</h2>
<section id="make-sure-newhybrids-is-installed-in-directory-quercus_bicolordatanewhybrids" class="level4">
<h4 class="anchored" data-anchor-id="make-sure-newhybrids-is-installed-in-directory-quercus_bicolordatanewhybrids">Make sure NewHybrids is installed in directory (~/Quercus_bicolor/data/NewHybrids/)</h4>
</section>
<section id="convert-.datastacksdataset2populations.snps.vcf-to-newhybrids-file-using-pgdspider-gui-program-then-run" class="level4">
<h4 class="anchored" data-anchor-id="convert-.datastacksdataset2populations.snps.vcf-to-newhybrids-file-using-pgdspider-gui-program-then-run">Convert “./data/STACKS/dataset2/populations.snps.vcf” to NEWHYBRIDS file using PGDSPider GUI program, then run:</h4>
<p><code>{.bash eval= FALSE} cd ~/Quercus_bicolor/data/NewHybrids/ ./newhybrids-no-gui-linux.exe -d ./newhybrids_data.txt --no-gui --burn-in 50000 --num-sweeps 500000 --seeds 17 47</code></p>
</section>
<section id="plot-hybrid-assignments-in-r" class="level4">
<h4 class="anchored" data-anchor-id="plot-hybrid-assignments-in-r">Plot hybrid assignments in R</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>nhq<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">"./data/NewHybrids/aa-PofZ.csv"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>nq<span class="ot">&lt;-</span>nhq[,<span class="dv">3</span><span class="sc">:</span><span class="dv">8</span>] <span class="sc">%&gt;%</span> <span class="fu">t</span>()</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>po<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">"./data/popmap_dataset2_version2"</span>, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>pops<span class="ot">&lt;-</span>po<span class="sc">$</span>pop</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>names<span class="ot">&lt;-</span>po<span class="sc">$</span>individual</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"lightblue"</span>,<span class="st">"olivedrab"</span>, <span class="st">"tomato"</span>,<span class="st">"pink"</span>,<span class="st">"purple"</span>, <span class="st">"gold"</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the output to a PDF file</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"barplot_newhybrids.png"</span>, <span class="at">width=</span><span class="dv">2000</span>, <span class="at">height=</span><span class="dv">800</span>, <span class="at">res=</span><span class="dv">150</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span> <span class="fu">c</span>(<span class="dv">11</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>)) </span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mgp =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">0</span>, <span class="sc">-</span><span class="dv">1</span>)) </span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">family=</span><span class="st">"mono"</span>)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the bar plot</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>plot <span class="ot">&lt;-</span> <span class="fu">barplot</span>(<span class="at">height=</span> nq, <span class="at">col=</span>my.colors, <span class="at">main=</span> <span class="st">"NewHybrids Analysis, Dataset 2"</span>, <span class="at">las=</span><span class="dv">2</span>, </span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">cex.names=</span><span class="dv">1</span>, <span class="at">cex.axis =</span> <span class="dv">1</span>, <span class="at">cex.lab=</span><span class="fl">1.5</span>, <span class="at">cex.main=</span><span class="fl">2.5</span>, </span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>                <span class="at">font.lab=</span><span class="dv">2</span>, <span class="at">font.axis=</span><span class="dv">2</span>, <span class="at">names.arg=</span><span class="fu">rep</span>(<span class="st">""</span>, <span class="fu">length</span>(names)), <span class="at">border=</span><span class="cn">FALSE</span>, <span class="at">space=</span><span class="dv">0</span>) <span class="co"># Remove individual labels</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Find unique group positions and midpoint for labeling</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>unique_pops <span class="ot">&lt;-</span> <span class="fu">unique</span>(pops)</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>label_positions <span class="ot">&lt;-</span> <span class="fu">sapply</span>(unique_pops, <span class="cf">function</span>(pop) <span class="fu">mean</span>(plot[pops <span class="sc">==</span> pop])) <span class="co"># Compute midpoints correctly</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Add population labels at midpoints</span></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at=</span>label_positions, <span class="at">labels=</span>unique_pops, <span class="at">tick=</span><span class="cn">FALSE</span>, <span class="at">cex.axis=</span><span class="fl">0.75</span>, <span class="at">las=</span><span class="dv">2</span>)</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Add vertical lines between groups</span></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>group_positions <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">diff</span>(<span class="fu">as.numeric</span>(<span class="fu">factor</span>(pops, <span class="at">levels=</span><span class="fu">unique</span>(pops)))) <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (pos <span class="cf">in</span> group_positions) {</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">abline</span>(<span class="at">v =</span> plot[pos] <span class="sc">+</span> <span class="fu">diff</span>(plot[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">/</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">lwd =</span> <span class="fl">1.5</span>)</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Add legend outside at the bottom</span></span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"right"</span>, <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"Quercus bicolor"</span>, <span class="st">"Quercus lyrata"</span>, <span class="st">"F1"</span>, <span class="st">"F2"</span>, </span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>                           <span class="st">"F1 x Quercus bicolor"</span>, <span class="st">"F1 x Quercus lyrata"</span>), </span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> my.colors, <span class="at">border =</span> <span class="st">"black"</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">inset=</span><span class="fl">0.15</span> ,  <span class="at">xpd =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="iii.-assess-heterozygosity-fst-fis-relative-to-q.-bicolor-ancestry-proportion-v4" class="level1">
<h1>III. Assess heterozygosity, Fst, Fis relative to Q. bicolor ancestry proportion (V4)</h1>
<section id="get-stacks-output-for-just-the-qbicolor-samples-exclude-reference-samples-and-those-with-higher-missing-data" class="level4">
<h4 class="anchored" data-anchor-id="get-stacks-output-for-just-the-qbicolor-samples-exclude-reference-samples-and-those-with-higher-missing-data">get STACKS output for just the Qbicolor samples (exclude reference samples and those with higher missing data)</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/data/STACKS/dataset1/"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>vc<span class="ot">&lt;-</span><span class="fu">read.vcfR</span>(<span class="st">"populations.snps.vcf"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">vcfR2genlight</span>(vc)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">gl.compliance.check</span>(g)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co">#filter to only include SNPs with callrate of 1 (can also be done within the STACKS populations module)</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">gl.filter.callrate</span>(g, <span class="at">threshold=</span><span class="dv">1</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>pop<span class="ot">&lt;-</span>pops<span class="sc">$</span>pop[<span class="dv">1</span><span class="sc">:</span><span class="dv">135</span>] <span class="sc">%&gt;%</span> <span class="fu">as.factor</span>()</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>strata<span class="ot">&lt;-</span>pops<span class="sc">$</span>pop2[<span class="dv">1</span><span class="sc">:</span><span class="dv">135</span>] <span class="sc">%&gt;%</span> as.data.frame</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>q<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">"qscores_k5.csv"</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>div<span class="ot">&lt;-</span><span class="fu">gl.report.diversity</span>(g)</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>heto<span class="ot">&lt;-</span><span class="fu">gl.report.heterozygosity</span>(g)</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert div$zero_D_alpha (named vector) into a data frame</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>div_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pop =</span> <span class="fu">names</span>(div<span class="sc">$</span>zero_D_alpha), <span class="at">D_alpha =</span> <span class="fu">as.numeric</span>(div<span class="sc">$</span>zero_D_alpha))</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge the two tables based on the "pop" column</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>merged_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(div_df, heto, <span class="at">by =</span> <span class="st">"pop"</span>)</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="co">#get Fst values and pairwise FST means</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>fst<span class="ot">&lt;-</span><span class="fu">gl.fst.pop</span>(g)</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>fst<span class="ot">&lt;-</span>fst <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>fst_no_diag <span class="ot">&lt;-</span> fst</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(fst_no_diag) <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an empty vector to store the means for each population</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>pairwise_means <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">"numeric"</span>, <span class="at">length =</span> <span class="fu">nrow</span>(fst))</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each population and calculate the pairwise average Fst</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(fst)) {</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get the row (excluding the diagonal value for this population)</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>  row_values <span class="ot">&lt;-</span> fst_no_diag[i, ]</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get the column (excluding the diagonal value for this population)</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>  col_values <span class="ot">&lt;-</span> fst_no_diag[, i]</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Combine both row and column values (without double-counting)</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>  combined_values <span class="ot">&lt;-</span> <span class="fu">c</span>(row_values, col_values)</span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove NA values and calculate the mean</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>  combined_values <span class="ot">&lt;-</span> combined_values[<span class="sc">!</span><span class="fu">is.na</span>(combined_values)]</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>  pairwise_means[i] <span class="ot">&lt;-</span> <span class="fu">mean</span>(combined_values)</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the results back to the matrix for easy reference</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pairwise_means) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(fst)</span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>pairwise_means<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(pairwise_means, <span class="at">row.names =</span> <span class="cn">NULL</span>)</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>pairwise_means<span class="sc">$</span>pop<span class="ot">&lt;-</span><span class="fu">rownames</span>(pairwise_means)</span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>pairwise_means</span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a><span class="co">#combine with het and div table</span></span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a>merged_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(merged_data, pairwise_means, <span class="at">by =</span> <span class="st">"pop"</span>)</span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a><span class="co">#add a column delineating core and edge pops</span></span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a>merged_data<span class="sc">$</span>pop3<span class="ot">&lt;-</span><span class="fu">substr</span>(merged_data<span class="sc">$</span>pop, <span class="at">start =</span> <span class="dv">1</span>, <span class="at">stop =</span> <span class="dv">4</span>)</span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a><span class="co">#add q scores averaged for each pop</span></span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>q_135<span class="ot">&lt;-</span>q[<span class="dv">1</span><span class="sc">:</span><span class="dv">135</span>,]</span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a>qavg <span class="ot">&lt;-</span> q_135 <span class="sc">%&gt;%</span></span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pop) <span class="sc">%&gt;%</span></span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(V4, mean, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="co"># Compute mean for numeric columns</span></span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a>merged_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(merged_data, qavg, <span class="at">by =</span> <span class="st">"pop"</span>)</span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a><span class="co">#regression of FIS vs ancestry proportion</span></span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a><span class="co"># filter out pops with less than 3 individuals</span></span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>merged_data_fil<span class="ot">&lt;-</span>merged_data[merged_data<span class="sc">$</span>n.Ind<span class="sc">&gt;</span><span class="dv">2</span>,]</span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a><span class="co">#transform FIS to be between 0 and 1 </span></span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a>merged_data_fil<span class="sc">$</span>FIStransf<span class="ot">&lt;-</span>((merged_data_fil<span class="sc">$</span>FIS<span class="sc">+</span><span class="dv">1</span>)<span class="sc">/</span><span class="dv">2</span>)</span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(betareg)</span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> betareg<span class="sc">::</span><span class="fu">betareg</span>(FIStransf <span class="sc">~</span> V4, <span class="at">data =</span> merged_data_fil, <span class="at">link =</span> <span class="st">"logit"</span>)</span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract design matrix (n × p)</span></span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a>s<span class="ot">&lt;-</span><span class="fu">summary</span>(mod1)</span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>s<span class="sc">$</span>coefficients<span class="sc">$</span>mean[<span class="dv">2</span>,<span class="dv">4</span>]</span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract R-squared and p-value</span></span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>r2_value <span class="ot">&lt;-</span> <span class="fu">summary</span>(mod1)<span class="sc">$</span>pseudo.r.squared</span>
<span id="cb15-66"><a href="#cb15-66" aria-hidden="true" tabindex="-1"></a>p_value <span class="ot">&lt;-</span> <span class="fu">summary</span>(mod1)<span class="sc">$</span>coefficients<span class="sc">$</span>mean[<span class="dv">2</span>,<span class="dv">4</span>]  <span class="co"># P-value for V4</span></span>
<span id="cb15-67"><a href="#cb15-67" aria-hidden="true" tabindex="-1"></a><span class="co"># Create text label for plot</span></span>
<span id="cb15-68"><a href="#cb15-68" aria-hidden="true" tabindex="-1"></a>lm_text <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Pseudo R² = "</span>, <span class="fu">round</span>(r2_value, <span class="dv">3</span>), </span>
<span id="cb15-69"><a href="#cb15-69" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"</span><span class="sc">\n</span><span class="st">P-value = "</span>, <span class="fu">format.pval</span>(p_value, <span class="at">digits =</span> <span class="dv">1</span>, <span class="at">eps =</span> <span class="fl">0.001</span>))</span>
<span id="cb15-70"><a href="#cb15-70" aria-hidden="true" tabindex="-1"></a><span class="co"># Create scatter plot with regression line and stats annotation</span></span>
<span id="cb15-71"><a href="#cb15-71" aria-hidden="true" tabindex="-1"></a>new_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">V4 =</span> <span class="fu">seq</span>(<span class="fu">min</span>(merged_data_fil<span class="sc">$</span>V4),<span class="fu">max</span>(merged_data_fil<span class="sc">$</span>V4), <span class="at">length.out =</span> <span class="dv">100</span>))</span>
<span id="cb15-72"><a href="#cb15-72" aria-hidden="true" tabindex="-1"></a><span class="co"># Predict values and back-transform them to original scale</span></span>
<span id="cb15-73"><a href="#cb15-73" aria-hidden="true" tabindex="-1"></a>new_data<span class="sc">$</span>predicted_transf <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod1, <span class="at">newdata =</span> new_data, <span class="at">type =</span> <span class="st">"response"</span>)</span>
<span id="cb15-74"><a href="#cb15-74" aria-hidden="true" tabindex="-1"></a>new_data<span class="sc">$</span>predicted_original <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> new_data<span class="sc">$</span>predicted_transf <span class="sc">-</span> <span class="dv">1</span>  <span class="co"># Back-transformation</span></span>
<span id="cb15-75"><a href="#cb15-75" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">family=</span><span class="st">"mono"</span>)</span>
<span id="cb15-76"><a href="#cb15-76" aria-hidden="true" tabindex="-1"></a>plot<span class="ot">&lt;-</span><span class="fu">ggplot</span>(merged_data_fil, <span class="fu">aes</span>(<span class="at">x =</span> V4, <span class="at">y =</span> FIS)) <span class="sc">+</span></span>
<span id="cb15-77"><a href="#cb15-77" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop)) <span class="sc">+</span>  <span class="co"># Color points by "pop"</span></span>
<span id="cb15-78"><a href="#cb15-78" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data=</span>new_data, <span class="at">color=</span><span class="st">'black'</span>,<span class="at">show.legend=</span><span class="cn">FALSE</span>,<span class="at">size=</span><span class="fl">0.5</span>,<span class="fu">aes</span>(<span class="at">x=</span>V4, <span class="at">y =</span> predicted_original, <span class="at">linetype =</span> <span class="st">"logit"</span>)) <span class="sc">+</span></span>
<span id="cb15-79"><a href="#cb15-79" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"sNMF Ancestry Proportion"</span>) <span class="sc">+</span></span>
<span id="cb15-80"><a href="#cb15-80" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="st">"F"</span>[IS])) <span class="sc">+</span></span>
<span id="cb15-81"><a href="#cb15-81" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb15-82"><a href="#cb15-82" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Inbreeding Coefficient as Function of Ancestry Proportion"</span>, <span class="at">color =</span> <span class="st">"Population"</span>) <span class="sc">+</span></span>
<span id="cb15-83"><a href="#cb15-83" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>, <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color=</span><span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb15-84"><a href="#cb15-84" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),  <span class="co"># Lighter major grid lines</span></span>
<span id="cb15-85"><a href="#cb15-85" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="fl">0.05</span>))<span class="sc">+</span>  </span>
<span id="cb15-86"><a href="#cb15-86" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, </span>
<span id="cb15-87"><a href="#cb15-87" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="fu">max</span>(merged_data_fil<span class="sc">$</span>V4) <span class="sc">*</span> <span class="fl">0.5</span>, </span>
<span id="cb15-88"><a href="#cb15-88" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> <span class="sc">-</span><span class="fl">0.1</span>, </span>
<span id="cb15-89"><a href="#cb15-89" aria-hidden="true" tabindex="-1"></a>           <span class="at">label =</span> lm_text, </span>
<span id="cb15-90"><a href="#cb15-90" aria-hidden="true" tabindex="-1"></a>           <span class="at">size =</span> <span class="dv">3</span>, <span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color=</span><span class="st">"black"</span>)</span>
<span id="cb15-91"><a href="#cb15-91" aria-hidden="true" tabindex="-1"></a><span class="er">)</span></span>
<span id="cb15-92"><a href="#cb15-92" aria-hidden="true" tabindex="-1"></a>plot</span>
<span id="cb15-93"><a href="#cb15-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-94"><a href="#cb15-94" aria-hidden="true" tabindex="-1"></a><span class="co">#regression of average pairwise Fst values</span></span>
<span id="cb15-95"><a href="#cb15-95" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> betareg<span class="sc">::</span><span class="fu">betareg</span>(pairwise_means <span class="sc">~</span> V4, <span class="at">data =</span> merged_data_fil, <span class="at">link =</span> <span class="st">"logit"</span>)</span>
<span id="cb15-96"><a href="#cb15-96" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract design matrix (n × p)</span></span>
<span id="cb15-97"><a href="#cb15-97" aria-hidden="true" tabindex="-1"></a>s<span class="ot">&lt;-</span><span class="fu">summary</span>(mod)</span>
<span id="cb15-98"><a href="#cb15-98" aria-hidden="true" tabindex="-1"></a>s<span class="sc">$</span>coefficients<span class="sc">$</span>mean[<span class="dv">2</span>,<span class="dv">4</span>]</span>
<span id="cb15-99"><a href="#cb15-99" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract R-squared and p-value</span></span>
<span id="cb15-100"><a href="#cb15-100" aria-hidden="true" tabindex="-1"></a>r2_value <span class="ot">&lt;-</span> <span class="fu">summary</span>(mod)<span class="sc">$</span>pseudo.r.squared</span>
<span id="cb15-101"><a href="#cb15-101" aria-hidden="true" tabindex="-1"></a>p_value <span class="ot">&lt;-</span> <span class="fu">summary</span>(mod)<span class="sc">$</span>coefficients<span class="sc">$</span>mean[<span class="dv">2</span>,<span class="dv">4</span>]  <span class="co"># P-value for V4</span></span>
<span id="cb15-102"><a href="#cb15-102" aria-hidden="true" tabindex="-1"></a><span class="co"># Create text label for plot</span></span>
<span id="cb15-103"><a href="#cb15-103" aria-hidden="true" tabindex="-1"></a>lm_text <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Pseudo R² = "</span>, <span class="fu">round</span>(r2_value, <span class="dv">3</span>), </span>
<span id="cb15-104"><a href="#cb15-104" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"</span><span class="sc">\n</span><span class="st">P-value = "</span>, <span class="fu">format.pval</span>(p_value, <span class="at">digits =</span> <span class="dv">3</span>, <span class="at">eps =</span> <span class="fl">0.001</span>))</span>
<span id="cb15-105"><a href="#cb15-105" aria-hidden="true" tabindex="-1"></a><span class="co"># Create scatter plot with regression line and stats annotation</span></span>
<span id="cb15-106"><a href="#cb15-106" aria-hidden="true" tabindex="-1"></a>plot<span class="ot">&lt;-</span><span class="fu">ggplot</span>(merged_data_fil, <span class="fu">aes</span>(<span class="at">x =</span> V4, <span class="at">y =</span> pairwise_means)) <span class="sc">+</span></span>
<span id="cb15-107"><a href="#cb15-107" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop)) <span class="sc">+</span>  <span class="co"># Color points by "pop"</span></span>
<span id="cb15-108"><a href="#cb15-108" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color=</span><span class="st">'black'</span>,<span class="at">show.legend=</span><span class="cn">FALSE</span>,<span class="at">size=</span><span class="fl">0.5</span>,<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">predict</span>(mod, merged_data_fil), <span class="at">linetype =</span> <span class="st">"logit"</span>)) <span class="sc">+</span></span>
<span id="cb15-109"><a href="#cb15-109" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"sNMF Ancestry Proportion"</span>) <span class="sc">+</span></span>
<span id="cb15-110"><a href="#cb15-110" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="st">"F"</span>[ST])) <span class="sc">+</span></span>
<span id="cb15-111"><a href="#cb15-111" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb15-112"><a href="#cb15-112" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Fixation Index as Function of Ancestry Proportion"</span>, <span class="at">color =</span> <span class="st">"Population"</span>) <span class="sc">+</span></span>
<span id="cb15-113"><a href="#cb15-113" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>, <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color=</span><span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb15-114"><a href="#cb15-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),  <span class="co"># Lighter major grid lines</span></span>
<span id="cb15-115"><a href="#cb15-115" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="fl">0.05</span>))<span class="sc">+</span>  </span>
<span id="cb15-116"><a href="#cb15-116" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, </span>
<span id="cb15-117"><a href="#cb15-117" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="fu">max</span>(merged_data_fil<span class="sc">$</span>V4) <span class="sc">*</span> <span class="fl">0.4</span>, </span>
<span id="cb15-118"><a href="#cb15-118" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> <span class="fu">max</span>(merged_data_fil<span class="sc">$</span>pairwise_means) <span class="sc">*</span> <span class="fl">0.6</span>, </span>
<span id="cb15-119"><a href="#cb15-119" aria-hidden="true" tabindex="-1"></a>           <span class="at">label =</span> lm_text, </span>
<span id="cb15-120"><a href="#cb15-120" aria-hidden="true" tabindex="-1"></a>           <span class="at">size =</span> <span class="dv">3</span>, <span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color=</span><span class="st">"black"</span>)</span>
<span id="cb15-121"><a href="#cb15-121" aria-hidden="true" tabindex="-1"></a><span class="er">)</span></span>
<span id="cb15-122"><a href="#cb15-122" aria-hidden="true" tabindex="-1"></a>plot</span>
<span id="cb15-123"><a href="#cb15-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-124"><a href="#cb15-124" aria-hidden="true" tabindex="-1"></a><span class="co">#also get by individual observed heterozygosity</span></span>
<span id="cb15-125"><a href="#cb15-125" aria-hidden="true" tabindex="-1"></a>het<span class="ot">&lt;-</span><span class="fu">gl.report.heterozygosity</span>(g, <span class="at">method=</span><span class="st">'ind'</span>)</span>
<span id="cb15-126"><a href="#cb15-126" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(het)[<span class="fu">colnames</span>(het) <span class="sc">==</span> <span class="st">"ind.name"</span>] <span class="ot">&lt;-</span> <span class="st">"names"</span></span>
<span id="cb15-127"><a href="#cb15-127" aria-hidden="true" tabindex="-1"></a>merged_data_ind <span class="ot">&lt;-</span> <span class="fu">merge</span>(het, q_135, <span class="at">by =</span> <span class="st">"names"</span>)</span>
<span id="cb15-128"><a href="#cb15-128" aria-hidden="true" tabindex="-1"></a>merged_data_ind<span class="sc">$</span>HetSitesHo<span class="ot">&lt;-</span>(merged_data_ind<span class="sc">$</span>Ho)<span class="sc">*</span><span class="dv">7955</span></span>
<span id="cb15-129"><a href="#cb15-129" aria-hidden="true" tabindex="-1"></a>merged_data_ind<span class="sc">$</span>HomSitesHo<span class="ot">&lt;-</span>(<span class="dv">1</span><span class="sc">-</span>merged_data_ind<span class="sc">$</span>Ho)<span class="sc">*</span><span class="dv">7955</span></span>
<span id="cb15-130"><a href="#cb15-130" aria-hidden="true" tabindex="-1"></a><span class="co">#binomial GLM model</span></span>
<span id="cb15-131"><a href="#cb15-131" aria-hidden="true" tabindex="-1"></a>modmod<span class="ot">&lt;-</span><span class="fu">glm</span>(<span class="fu">cbind</span>(HetSitesHo,HomSitesHo)<span class="sc">~</span>V4, <span class="at">data=</span>merged_data_ind, <span class="at">family=</span><span class="fu">binomial</span>(<span class="at">link=</span><span class="st">'logit'</span>))</span>
<span id="cb15-132"><a href="#cb15-132" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(modmod)</span>
<span id="cb15-133"><a href="#cb15-133" aria-hidden="true" tabindex="-1"></a>predictedmodel<span class="ot">&lt;-</span><span class="fu">predict.glm</span>(modmod,merged_data_ind,<span class="at">se.fit =</span> T)</span>
<span id="cb15-134"><a href="#cb15-134" aria-hidden="true" tabindex="-1"></a>ci_lwr <span class="ot">&lt;-</span> <span class="fu">with</span>(predictedmodel, <span class="fu">plogis</span>(fit <span class="sc">-</span> <span class="fl">1.96</span><span class="sc">*</span>se.fit))</span>
<span id="cb15-135"><a href="#cb15-135" aria-hidden="true" tabindex="-1"></a>ci_upr <span class="ot">&lt;-</span> <span class="fu">with</span>(predictedmodel, <span class="fu">plogis</span>(fit <span class="sc">+</span> <span class="fl">1.96</span><span class="sc">*</span>se.fit))</span>
<span id="cb15-136"><a href="#cb15-136" aria-hidden="true" tabindex="-1"></a>merged_data_ind<span class="ot">&lt;-</span><span class="fu">cbind</span>(merged_data_ind,predictedmodel)</span>
<span id="cb15-137"><a href="#cb15-137" aria-hidden="true" tabindex="-1"></a>merged_data_ind<span class="sc">$</span>fit2<span class="ot">&lt;-</span><span class="fu">plogis</span>(merged_data_ind<span class="sc">$</span>fit)</span>
<span id="cb15-138"><a href="#cb15-138" aria-hidden="true" tabindex="-1"></a>merged_data_ind<span class="sc">$</span>lwr<span class="ot">&lt;-</span>ci_lwr</span>
<span id="cb15-139"><a href="#cb15-139" aria-hidden="true" tabindex="-1"></a>merged_data_ind<span class="sc">$</span>upr<span class="ot">&lt;-</span>ci_upr</span>
<span id="cb15-140"><a href="#cb15-140" aria-hidden="true" tabindex="-1"></a><span class="co">#pseudo R2</span></span>
<span id="cb15-141"><a href="#cb15-141" aria-hidden="true" tabindex="-1"></a>r2_value<span class="ot">&lt;-</span><span class="dv">1</span><span class="sc">-</span>modmod<span class="sc">$</span>deviance<span class="sc">/</span>modmod<span class="sc">$</span>null.deviance</span>
<span id="cb15-142"><a href="#cb15-142" aria-hidden="true" tabindex="-1"></a><span class="co">#get pvalue</span></span>
<span id="cb15-143"><a href="#cb15-143" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stats)</span>
<span id="cb15-144"><a href="#cb15-144" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(modmod, <span class="at">test=</span><span class="st">"Chisq"</span>)</span>
<span id="cb15-145"><a href="#cb15-145" aria-hidden="true" tabindex="-1"></a>lm_text <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Pseudo R² = "</span>, <span class="fu">round</span>(r2_value, <span class="dv">3</span>),</span>
<span id="cb15-146"><a href="#cb15-146" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"</span><span class="sc">\n</span><span class="st">p-value = &lt;0.0001"</span>)</span>
<span id="cb15-147"><a href="#cb15-147" aria-hidden="true" tabindex="-1"></a>plot<span class="ot">&lt;-</span><span class="fu">ggplot</span>(merged_data_ind, <span class="fu">aes</span>(<span class="at">x =</span> V4, <span class="at">y =</span> Ho)) <span class="sc">+</span></span>
<span id="cb15-148"><a href="#cb15-148" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop)) <span class="sc">+</span>  <span class="co"># Color points by "pop"</span></span>
<span id="cb15-149"><a href="#cb15-149" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> lwr, <span class="at">ymax =</span> upr), <span class="at">fill =</span> <span class="st">"grey"</span>, <span class="at">alpha =</span> <span class="fl">0.15</span>) <span class="sc">+</span>  <span class="co"># Neutral-colored error ribbon</span></span>
<span id="cb15-150"><a href="#cb15-150" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> fit2), <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> .<span class="dv">5</span>) <span class="sc">+</span>  <span class="co"># Regression line</span></span>
<span id="cb15-151"><a href="#cb15-151" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"sNMF Ancestry Proportion"</span>) <span class="sc">+</span></span>
<span id="cb15-152"><a href="#cb15-152" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="st">'H'</span>[O])) <span class="sc">+</span></span>
<span id="cb15-153"><a href="#cb15-153" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb15-154"><a href="#cb15-154" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Observed Heterozygosity as Function of Ancestry Proportion"</span>, <span class="at">color =</span> <span class="st">"Population"</span>) <span class="sc">+</span></span>
<span id="cb15-155"><a href="#cb15-155" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>, <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color=</span><span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb15-156"><a href="#cb15-156" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray85"</span>, <span class="at">size =</span> <span class="fl">0.05</span>),  <span class="co"># Lighter major grid lines</span></span>
<span id="cb15-157"><a href="#cb15-157" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"gray92"</span>, <span class="at">size =</span> <span class="fl">0.05</span>))<span class="sc">+</span>  </span>
<span id="cb15-158"><a href="#cb15-158" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, </span>
<span id="cb15-159"><a href="#cb15-159" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="fu">max</span>(merged_data_ind<span class="sc">$</span>V4) <span class="sc">*</span> <span class="fl">0.5</span>, </span>
<span id="cb15-160"><a href="#cb15-160" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> <span class="fu">max</span>(merged_data_ind<span class="sc">$</span>Ho) <span class="sc">*</span> <span class="fl">0.9</span>, </span>
<span id="cb15-161"><a href="#cb15-161" aria-hidden="true" tabindex="-1"></a>           <span class="at">label =</span> lm_text, </span>
<span id="cb15-162"><a href="#cb15-162" aria-hidden="true" tabindex="-1"></a>           <span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">family =</span> <span class="st">"mono"</span>, <span class="at">color=</span><span class="st">"black"</span>)</span>
<span id="cb15-163"><a href="#cb15-163" aria-hidden="true" tabindex="-1"></a><span class="er">)</span></span>
<span id="cb15-164"><a href="#cb15-164" aria-hidden="true" tabindex="-1"></a>plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="iv.-assess-gd-ibd-amova-and-structure-only-including-pure-q.-bicolor" class="level1">
<h1>IV. Assess GD, IBD, AMOVA, and structure only including “pure” Q. bicolor</h1>
<section id="prepare-data" class="level4">
<h4 class="anchored" data-anchor-id="prepare-data">Prepare data</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>vc<span class="ot">&lt;-</span><span class="fu">read.vcfR</span>(<span class="st">"./data/STACKS/dataset3/populations.snps.vcf"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">vcfR2genlight</span>(vc)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">gl.compliance.check</span>(g)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>names<span class="ot">&lt;-</span>g<span class="sc">@</span>ind.names</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>pop<span class="ot">&lt;-</span>pure<span class="sc">$</span>pop <span class="sc">%&gt;%</span> <span class="fu">as.factor</span>()</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>strata<span class="ot">&lt;-</span>pure<span class="sc">$</span>pop2 <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>strata<span class="ot">&lt;-</span>strata</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>coords<span class="ot">&lt;-</span><span class="fu">cbind</span>(pure<span class="sc">$</span>LAT,pure<span class="sc">$</span>LON) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>other<span class="sc">$</span>latlon<span class="ot">&lt;-</span>coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-ibd-on-edge-and-core-together" class="level4">
<h4 class="anchored" data-anchor-id="run-ibd-on-edge-and-core-together">Run IBD on edge and core together</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ibd<span class="ot">&lt;-</span><span class="fu">gl.ibd</span>(g, <span class="at">coordinates=</span><span class="st">"latlon"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>ibd</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co">#separate edge and core</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>edge<span class="ot">&lt;-</span>g[g<span class="sc">@</span>strata<span class="sc">$</span>.<span class="sc">==</span><span class="st">'EDGE'</span>]</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>core<span class="ot">&lt;-</span>g[g<span class="sc">@</span>strata<span class="sc">$</span>.<span class="sc">==</span><span class="st">'CORE'</span>]</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co">#run IBD on each separately</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>ibd_edge<span class="ot">&lt;-</span><span class="fu">gl.ibd</span>(edge, <span class="at">coordinates=</span><span class="st">"latlon"</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>ibd_core<span class="ot">&lt;-</span><span class="fu">gl.ibd</span>(core, <span class="at">coordinates=</span><span class="st">"latlon"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-amova" class="level4">
<h4 class="anchored" data-anchor-id="run-amova">Run AMOVA</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>amova<span class="ot">&lt;-</span><span class="fu">gl.amova</span>(g)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>amova_edge<span class="ot">&lt;-</span><span class="fu">gl.amova</span>(edge)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>amova_core<span class="ot">&lt;-</span><span class="fu">gl.amova</span>(core)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-snmf" class="level4">
<h4 class="anchored" data-anchor-id="run-snmf">Run sNMF</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl2geno</span>(g, <span class="at">outfile=</span> <span class="st">"g_geno"</span>,<span class="at">outpath=</span> <span class="fu">getwd</span>())</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>names<span class="ot">&lt;-</span>g<span class="sc">@</span>ind.names</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co">#detect admixture</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>project <span class="ot">=</span> <span class="fu">snmf</span>(<span class="st">"g_geno.geno"</span>, <span class="at">K =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>, <span class="at">entropy =</span> <span class="cn">TRUE</span>, <span class="at">repetitions =</span> <span class="dv">100</span>, <span class="at">seed=</span><span class="dv">42</span>, <span class="at">project =</span> <span class="st">"new"</span>, <span class="at">iterations=</span><span class="dv">1000000</span>, <span class="at">alpha=</span><span class="dv">10</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>project <span class="ot">=</span> <span class="fu">load.snmfProject</span>(<span class="st">"g_geno.snmfProject"</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(project, <span class="at">col =</span> <span class="st">"blue"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">1.2</span>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co"># select the best run for K = 4 clusters </span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>best <span class="ot">=</span> <span class="fu">which.min</span>(<span class="fu">cross.entropy</span>(project, <span class="at">K =</span> <span class="dv">5</span>)) </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"tomato"</span>, <span class="st">"olivedrab"</span>, <span class="st">"gold"</span>,<span class="st">"lightblue"</span>,<span class="st">"pink"</span>,<span class="st">"purple"</span>,<span class="st">"orange"</span>) </span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="fu">windows</span>()</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="fu">barchart</span>(project, <span class="at">K =</span> <span class="dv">5</span>, <span class="at">run =</span> best, <span class="at">border =</span> <span class="cn">NA</span>,<span class="at">sort.by.Q=</span><span class="cn">FALSE</span>, <span class="at">space =</span> <span class="dv">0</span>, <span class="at">col =</span> my.colors, <span class="at">xlab =</span> <span class="st">"Individuals"</span>, <span class="at">ylab =</span> <span class="st">"Ancestry proportions"</span>, <span class="at">main =</span> <span class="st">"Ancestry matrix"</span>)<span class="ot">-&gt;</span> bp </span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(names), <span class="at">labels =</span> names, <span class="at">las=</span><span class="dv">2</span>, <span class="at">cex.axis =</span> .<span class="dv">5</span>)</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>q_scores_pure<span class="ot">&lt;-</span><span class="fu">Q</span>(project, <span class="at">K =</span> <span class="dv">5</span>, <span class="fu">which.min</span>(<span class="fu">cross.entropy</span>(project, <span class="at">K =</span> <span class="dv">5</span>)))</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>q_pure<span class="ot">&lt;-</span><span class="fu">cbind</span>(pure,q_scores_pure) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(q_pure,<span class="st">"qscores_k5_pure.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="get-pop-gen-stats-on-dataset-3" class="level4">
<h4 class="anchored" data-anchor-id="get-pop-gen-stats-on-dataset-3">Get pop gen stats on Dataset 3</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get ttests on het and fis fst values</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>het<span class="ot">&lt;-</span><span class="fu">gl.report.heterozygosity</span>(g)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>div<span class="ot">&lt;-</span><span class="fu">gl.report.diversity</span>(g)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert div$zero_D_alpha (named vector) into a data frame</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>div_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pop =</span> <span class="fu">names</span>(div<span class="sc">$</span>zero_D_alpha), <span class="at">D_alpha =</span> <span class="fu">as.numeric</span>(div<span class="sc">$</span>zero_D_alpha))</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge the two tables based on the "pop" column</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>merged_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(div_df, het, <span class="at">by =</span> <span class="st">"pop"</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co">#get fst values</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>fst<span class="ot">&lt;-</span><span class="fu">gl.fst.pop</span>(g)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>fst_no_diag <span class="ot">&lt;-</span> fst</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(fst_no_diag) <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an empty vector to store the means for each population</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>pairwise_means <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">"numeric"</span>, <span class="at">length =</span> <span class="fu">nrow</span>(fst))</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each population and calculate the pairwise average Fst</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(fst)) {</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get the row (excluding the diagonal value for this population)</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  row_values <span class="ot">&lt;-</span> fst_no_diag[i, ]</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get the column (excluding the diagonal value for this population)</span></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>  col_values <span class="ot">&lt;-</span> fst_no_diag[, i]</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Combine both row and column values (without double-counting)</span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>  combined_values <span class="ot">&lt;-</span> <span class="fu">c</span>(row_values, col_values)</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove NA values and calculate the mean</span></span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>  combined_values <span class="ot">&lt;-</span> combined_values[<span class="sc">!</span><span class="fu">is.na</span>(combined_values)]</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>  pairwise_means[i] <span class="ot">&lt;-</span> <span class="fu">mean</span>(combined_values)</span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the results back to the matrix for easy reference</span></span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pairwise_means) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(fst)</span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>pairwise_means<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(pairwise_means, <span class="at">row.names =</span> <span class="cn">NULL</span>)</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a>pairwise_means<span class="sc">$</span>pop<span class="ot">&lt;-</span><span class="fu">rownames</span>(pairwise_means)</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>pairwise_means</span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>merged_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(merged_data, pairwise_means, <span class="at">by =</span> <span class="st">"pop"</span>)</span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a>merged_data<span class="sc">$</span>pop2<span class="ot">&lt;-</span><span class="fu">substr</span>(merged_data<span class="sc">$</span>pop, <span class="at">start =</span> <span class="dv">1</span>, <span class="at">stop =</span> <span class="dv">4</span>)</span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a>merged_data</span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(merged_data)[<span class="fu">colnames</span>(merged_data)<span class="sc">==</span><span class="st">"pairwise_means"</span>]<span class="ot">&lt;-</span><span class="st">"Avg_Fst"</span></span>
<span id="cb20-37"><a href="#cb20-37" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(merged_data, <span class="st">"merged_data_PURE.csv"</span>)</span>
<span id="cb20-38"><a href="#cb20-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-39"><a href="#cb20-39" aria-hidden="true" tabindex="-1"></a>popedge<span class="ot">&lt;-</span>merged_data[merged_data<span class="sc">$</span>pop2<span class="sc">==</span><span class="st">'EDGE'</span>,]</span>
<span id="cb20-40"><a href="#cb20-40" aria-hidden="true" tabindex="-1"></a>popcore<span class="ot">&lt;-</span>merged_data[merged_data<span class="sc">$</span>pop2<span class="sc">==</span><span class="st">'CORE'</span>,]</span>
<span id="cb20-41"><a href="#cb20-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-42"><a href="#cb20-42" aria-hidden="true" tabindex="-1"></a>mergedfilt<span class="ot">&lt;-</span>merged_data[merged_data<span class="sc">$</span>nInd<span class="sc">&gt;</span><span class="dv">2</span>,]</span>
<span id="cb20-43"><a href="#cb20-43" aria-hidden="true" tabindex="-1"></a>popedge<span class="ot">&lt;-</span>mergedfilt[mergedfilt<span class="sc">$</span>pop2<span class="sc">==</span><span class="st">'EDGE'</span>,]</span>
<span id="cb20-44"><a href="#cb20-44" aria-hidden="true" tabindex="-1"></a>popcore<span class="ot">&lt;-</span>mergedfilt[mergedfilt<span class="sc">$</span>pop2<span class="sc">==</span><span class="st">'CORE'</span>,]</span>
<span id="cb20-45"><a href="#cb20-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-46"><a href="#cb20-46" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(popedge<span class="sc">$</span>Ho.adj, popcore<span class="sc">$</span>Ho.adj,<span class="at">alternative=</span><span class="st">'less'</span>, <span class="at">var.equal =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-47"><a href="#cb20-47" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(popedge<span class="sc">$</span>He.adj, popcore<span class="sc">$</span>He.adj,<span class="at">alternative=</span><span class="st">'less'</span>, <span class="at">var.equal =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-48"><a href="#cb20-48" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(popedge<span class="sc">$</span>FIS, popcore<span class="sc">$</span>FIS,<span class="at">alternative=</span><span class="st">'greater'</span>, <span class="at">var.equal =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-49"><a href="#cb20-49" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(popedge<span class="sc">$</span>Avg_Fst, popcore<span class="sc">$</span>Avg_Fst,<span class="at">alternative=</span><span class="st">'greater'</span>, <span class="at">var.equal =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-50"><a href="#cb20-50" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(popedge<span class="sc">$</span>D_alpha, popcore<span class="sc">$</span>D_alpha,<span class="at">alternative=</span><span class="st">'less'</span>, <span class="at">var.equal =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-51"><a href="#cb20-51" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(popedge<span class="sc">$</span>pi, popcore<span class="sc">$</span>pi,<span class="at">alternative=</span><span class="st">'less'</span>, <span class="at">var.equal =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-52"><a href="#cb20-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-53"><a href="#cb20-53" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(popedge<span class="sc">$</span>FIS)</span>
<span id="cb20-54"><a href="#cb20-54" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(popcore<span class="sc">$</span>FIS)</span>
<span id="cb20-55"><a href="#cb20-55" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(popedge<span class="sc">$</span>FIS)</span>
<span id="cb20-56"><a href="#cb20-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-57"><a href="#cb20-57" aria-hidden="true" tabindex="-1"></a><span class="fu">var.test</span>(popedge<span class="sc">$</span>Ho, popcore<span class="sc">$</span>Ho, <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb20-58"><a href="#cb20-58" aria-hidden="true" tabindex="-1"></a><span class="fu">var.test</span>(popedge<span class="sc">$</span>He, popcore<span class="sc">$</span>He, <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb20-59"><a href="#cb20-59" aria-hidden="true" tabindex="-1"></a><span class="fu">var.test</span>(popedge<span class="sc">$</span>FIS, popcore<span class="sc">$</span>FIS, <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb20-60"><a href="#cb20-60" aria-hidden="true" tabindex="-1"></a><span class="fu">var.test</span>(popedge<span class="sc">$</span>Avg_Fst, popcore<span class="sc">$</span>Avg_Fst, <span class="at">alternative=</span><span class="st">'greater'</span>)</span>
<span id="cb20-61"><a href="#cb20-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-62"><a href="#cb20-62" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span>
<span id="cb20-63"><a href="#cb20-63" aria-hidden="true" tabindex="-1"></a><span class="co">#test if variances are sig diff</span></span>
<span id="cb20-64"><a href="#cb20-64" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Group1"</span>, <span class="fu">length</span>(edge<span class="sc">$</span>FIS)), <span class="fu">rep</span>(<span class="st">"Group2"</span>, <span class="fu">length</span>(core<span class="sc">$</span>FIS)))</span>
<span id="cb20-65"><a href="#cb20-65" aria-hidden="true" tabindex="-1"></a>values <span class="ot">&lt;-</span> <span class="fu">c</span>(edge<span class="sc">$</span>FIS, core<span class="sc">$</span>FIS)</span>
<span id="cb20-66"><a href="#cb20-66" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(edge<span class="sc">$</span>FIS)</span>
<span id="cb20-67"><a href="#cb20-67" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(core<span class="sc">$</span>FIS)</span>
<span id="cb20-68"><a href="#cb20-68" aria-hidden="true" tabindex="-1"></a>ltest<span class="ot">&lt;-</span><span class="fu">leveneTest</span>(values <span class="sc">~</span> group)</span>
<span id="cb20-69"><a href="#cb20-69" aria-hidden="true" tabindex="-1"></a>ltest<span class="sc">$</span><span class="st">`</span><span class="at">Pr(&gt;F)</span><span class="st">`</span><span class="sc">/</span><span class="dv">2</span></span>
<span id="cb20-70"><a href="#cb20-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-71"><a href="#cb20-71" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Group1"</span>, <span class="fu">length</span>(edge<span class="sc">$</span>He)), <span class="fu">rep</span>(<span class="st">"Group2"</span>, <span class="fu">length</span>(core<span class="sc">$</span>He)))</span>
<span id="cb20-72"><a href="#cb20-72" aria-hidden="true" tabindex="-1"></a>values <span class="ot">&lt;-</span> <span class="fu">c</span>(edge<span class="sc">$</span>He, core<span class="sc">$</span>He)</span>
<span id="cb20-73"><a href="#cb20-73" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(edge<span class="sc">$</span>He)</span>
<span id="cb20-74"><a href="#cb20-74" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(core<span class="sc">$</span>He)</span>
<span id="cb20-75"><a href="#cb20-75" aria-hidden="true" tabindex="-1"></a>ltest<span class="ot">&lt;-</span><span class="fu">leveneTest</span>(values <span class="sc">~</span> group)</span>
<span id="cb20-76"><a href="#cb20-76" aria-hidden="true" tabindex="-1"></a>ltest<span class="sc">$</span><span class="st">`</span><span class="at">Pr(&gt;F)</span><span class="st">`</span><span class="sc">/</span><span class="dv">2</span></span>
<span id="cb20-77"><a href="#cb20-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-78"><a href="#cb20-78" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Group1"</span>, <span class="fu">length</span>(edge<span class="sc">$</span>Allelic.Richness)), <span class="fu">rep</span>(<span class="st">"Group2"</span>, <span class="fu">length</span>(core<span class="sc">$</span>Allelic.Richness)))</span>
<span id="cb20-79"><a href="#cb20-79" aria-hidden="true" tabindex="-1"></a>values <span class="ot">&lt;-</span> <span class="fu">c</span>(edge<span class="sc">$</span>Allelic.Richness, core<span class="sc">$</span>Allelic.Richness)</span>
<span id="cb20-80"><a href="#cb20-80" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(edge<span class="sc">$</span>Allelic.Richness)</span>
<span id="cb20-81"><a href="#cb20-81" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(core<span class="sc">$</span>Allelic.Richness)</span>
<span id="cb20-82"><a href="#cb20-82" aria-hidden="true" tabindex="-1"></a>ltest<span class="ot">&lt;-</span><span class="fu">leveneTest</span>(values <span class="sc">~</span> group)</span>
<span id="cb20-83"><a href="#cb20-83" aria-hidden="true" tabindex="-1"></a>ltest<span class="sc">$</span><span class="st">`</span><span class="at">Pr(&gt;F)</span><span class="st">`</span><span class="sc">/</span><span class="dv">2</span></span>
<span id="cb20-84"><a href="#cb20-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-85"><a href="#cb20-85" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Group1"</span>, <span class="fu">length</span>(edge<span class="sc">$</span>pi)), <span class="fu">rep</span>(<span class="st">"Group2"</span>, <span class="fu">length</span>(core<span class="sc">$</span>pi)))</span>
<span id="cb20-86"><a href="#cb20-86" aria-hidden="true" tabindex="-1"></a>values <span class="ot">&lt;-</span> <span class="fu">c</span>(edge<span class="sc">$</span>pi, core<span class="sc">$</span>pi)</span>
<span id="cb20-87"><a href="#cb20-87" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(edge<span class="sc">$</span>pi)</span>
<span id="cb20-88"><a href="#cb20-88" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(core<span class="sc">$</span>pi)</span>
<span id="cb20-89"><a href="#cb20-89" aria-hidden="true" tabindex="-1"></a>ltest<span class="ot">&lt;-</span><span class="fu">leveneTest</span>(values <span class="sc">~</span> group)</span>
<span id="cb20-90"><a href="#cb20-90" aria-hidden="true" tabindex="-1"></a>ltest<span class="sc">$</span><span class="st">`</span><span class="at">Pr(&gt;F)</span><span class="st">`</span><span class="sc">/</span><span class="dv">2</span></span>
<span id="cb20-91"><a href="#cb20-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-92"><a href="#cb20-92" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Group1"</span>, <span class="fu">length</span>(popedge<span class="sc">$</span>pi)), <span class="fu">rep</span>(<span class="st">"Group2"</span>, <span class="fu">length</span>(popcore<span class="sc">$</span>pi)))</span>
<span id="cb20-93"><a href="#cb20-93" aria-hidden="true" tabindex="-1"></a>values <span class="ot">&lt;-</span> <span class="fu">c</span>(popedge<span class="sc">$</span>pi, popcore<span class="sc">$</span>pi)</span>
<span id="cb20-94"><a href="#cb20-94" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(popedge<span class="sc">$</span>pi)</span>
<span id="cb20-95"><a href="#cb20-95" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(popcore<span class="sc">$</span>pi)</span>
<span id="cb20-96"><a href="#cb20-96" aria-hidden="true" tabindex="-1"></a>ltest<span class="ot">&lt;-</span><span class="fu">leveneTest</span>(values <span class="sc">~</span> group)</span>
<span id="cb20-97"><a href="#cb20-97" aria-hidden="true" tabindex="-1"></a>ltest<span class="sc">$</span><span class="st">`</span><span class="at">Pr(&gt;F)</span><span class="st">`</span><span class="sc">/</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-treemix" class="level4">
<h4 class="anchored" data-anchor-id="run-treemix">Run Treemix</h4>
<section id="prepare-and-check-data-in-r" class="level5">
<h5 class="anchored" data-anchor-id="prepare-and-check-data-in-r">Prepare and check data in R</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>vc<span class="ot">&lt;-</span><span class="fu">read.vcfR</span>(<span class="st">"./data/STACKS/dataset2_NoMissing/populations.snps.vcf"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>g<span class="ot">&lt;-</span><span class="fu">vcfR2genlight</span>(vc)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>pm_treemix<span class="ot">&lt;-</span><span class="fu">read.delim</span>(<span class="st">"./data/popmap_treemix.txt"</span>,<span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>g<span class="sc">@</span>pop<span class="ot">&lt;-</span>pm_treemix<span class="sc">$</span>V2 <span class="sc">%&gt;%</span> <span class="fu">as.factor</span>()</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"./data/treemix"</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="fu">gl2treemix</span>(g, <span class="at">outfile=</span><span class="st">"g_141.treemix.gz"</span>, <span class="at">outpath=</span><span class="fu">getwd</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-treemix-in-command-line-make-sure-treemix-is-installed-in-quercus_bicolordatatreemix-directory" class="level5">
<h5 class="anchored" data-anchor-id="run-treemix-in-command-line-make-sure-treemix-is-installed-in-quercus_bicolordatatreemix-directory">Run treemix in command line (make sure treemix is installed in “~/Quercus_bicolor/data/treemix” directory)</h5>
<div class="sourceCode" id="cb22" data-eval="FALSE"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/Quercus_bicolor/data/treemix</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.1.1 <span class="at">-root</span> 3 <span class="at">-m</span> 1 <span class="at">-bootstrap</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.2.1 <span class="at">-root</span> 3 <span class="at">-m</span> 1 <span class="at">-bootstrap</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.3.1 <span class="at">-root</span> 3 <span class="at">-m</span> 1 <span class="at">-bootstrap</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.4.1 <span class="at">-root</span> 3 <span class="at">-m</span> 1 <span class="at">-bootstrap</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.5.1 <span class="at">-root</span> 3 <span class="at">-m</span> 1 <span class="at">-bootstrap</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.1.2 <span class="at">-root</span> 3 <span class="at">-m</span> 2 <span class="at">-bootstrap</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.2.2 <span class="at">-root</span> 3 <span class="at">-m</span> 2 <span class="at">-bootstrap</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.3.2 <span class="at">-root</span> 3 <span class="at">-m</span> 2 <span class="at">-bootstrap</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.4.2 <span class="at">-root</span> 3 <span class="at">-m</span> 2 <span class="at">-bootstrap</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.5.2 <span class="at">-root</span> 3 <span class="at">-m</span> 2 <span class="at">-bootstrap</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.1.3 <span class="at">-root</span> 3 <span class="at">-m</span> 3 <span class="at">-bootstrap</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.2.3 <span class="at">-root</span> 3 <span class="at">-m</span> 3 <span class="at">-bootstrap</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.3.3 <span class="at">-root</span> 3 <span class="at">-m</span> 3 <span class="at">-bootstrap</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.4.3 <span class="at">-root</span> 3 <span class="at">-m</span> 3 <span class="at">-bootstrap</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="ex">./treemix-1.13/treemix-1.13/src/treemix</span> <span class="at">-i</span> g_141.treemix.gz <span class="at">-o</span> ./treemix_out/out.5.3 <span class="at">-root</span> 3 <span class="at">-m</span> 3 <span class="at">-bootstrap</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="repeat-for-values-of-m-ranging-1-10-with-5-repetitions-following-same-pattern-of-file-names" class="level5">
<h5 class="anchored" data-anchor-id="repeat-for-values-of-m-ranging-1-10-with-5-repetitions-following-same-pattern-of-file-names">Repeat for values of m ranging 1-10 with 5 repetitions, following same pattern of file names</h5>
</section>
<section id="plot-treemix-results-in-r" class="level5">
<h5 class="anchored" data-anchor-id="plot-treemix-results-in-r">Plot treemix results in R</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"~/Quercus_bicolor/treemix-1.13/treemix-1.13/src/plotting_funcs.R"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/data/treemix/treemix_out/"</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="at">pattern=</span><span class="st">'</span><span class="sc">\\</span><span class="st">.modelcov.gz$'</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(OptM)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>optm<span class="ot">&lt;-</span><span class="fu">optM</span>(<span class="fu">getwd</span>())</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_optM</span>(optm)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="fu">windows</span>()</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_tree</span>(<span class="st">"~/Quercus_bicolor/data/treemix/treemix_out/out.1.6"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="run-tess3r" class="level4">
<h4 class="anchored" data-anchor-id="run-tess3r">Run Tess3r</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"bcm-uga/TESS3_encho_sen"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tess3r)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#impute missing values</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>best <span class="ot">=</span> <span class="fu">which.min</span>(<span class="fu">cross.entropy</span>(project, <span class="at">K =</span> <span class="dv">5</span>)) </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="fu">impute</span>(project, <span class="st">"gl_geno.lfmm"</span>, <span class="at">method =</span> <span class="st">'mode'</span>, <span class="at">K =</span> <span class="dv">5</span>, <span class="at">run =</span> best) </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>coordinates<span class="ot">&lt;-</span><span class="fu">as.matrix</span>(coords)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(coordinates)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="st">"gl_geno.lfmm"</span>)</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(maps)</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(coordinates, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> .<span class="dv">5</span>, </span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Longitude (°E)"</span>, <span class="at">ylab =</span> <span class="st">"Latitude (°N)"</span>)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="fu">map</span>(<span class="at">database=</span> <span class="st">"county"</span>, <span class="at">add =</span> T, <span class="at">interior =</span> T)</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>project<span class="ot">=</span> <span class="fu">load.lfmmProject</span>(<span class="st">"gl_geno.lfmmProject"</span>)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>lfmm<span class="ot">&lt;-</span><span class="fu">read.lfmm</span>(<span class="st">"glv_geno.lfmm_imputed.lfmm"</span>)</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>tess3.obj <span class="ot">&lt;-</span> <span class="fu">tess3</span>(<span class="at">X =</span> lfmm, <span class="at">coord =</span> coords, <span class="at">K =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>, </span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>                   <span class="at">method =</span> <span class="st">"projected.ls"</span>, <span class="at">ploidy =</span> <span class="dv">2</span>, <span class="at">openMP.core.num =</span> <span class="dv">4</span>, <span class="at">max.iteration =</span> <span class="dv">1000000</span>, <span class="at">rep=</span><span class="dv">10</span>) </span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tess3.obj, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="st">"blue"</span>,</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of ancestral populations"</span>,</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Cross-validation score"</span>)</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="co"># retrieve tess3 Q matrix for K = 5 clusters </span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>q.matrix <span class="ot">&lt;-</span> <span class="fu">qmatrix</span>(tess3.obj, <span class="at">K =</span> <span class="dv">4</span>)</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a><span class="co"># STRUCTURE-like barplot for the Q-matrix </span></span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>pal<span class="ot">&lt;-</span><span class="fu">CreatePalette</span>(<span class="at">color.vector =</span> <span class="fu">c</span>(<span class="st">"tomato"</span>, <span class="st">"chartreuse"</span>,<span class="st">"blue"</span>, <span class="st">"gold"</span>,</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"violet"</span>, <span class="st">"olivedrab"</span>, <span class="st">"purple"</span>, <span class="st">"brown"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>), <span class="at">palette.length =</span> <span class="dv">4</span>)</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(q.matrix, <span class="at">border =</span> <span class="cn">NA</span>, <span class="at">sort.by.Q=</span><span class="cn">FALSE</span>,<span class="at">space =</span> <span class="dv">0</span>,</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Individuals"</span>, <span class="at">ylab =</span> <span class="st">"Ancestry proportions"</span>, </span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Ancestry matrix"</span>) <span class="ot">-&gt;</span> bp</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>q<span class="ot">&lt;-</span>q.matrix <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="fu">cbind</span>(order)</span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(q.matrix), <span class="at">labels =</span> plotorder, <span class="at">las =</span> <span class="dv">3</span>, <span class="at">cex.axis =</span> .<span class="dv">4</span>) </span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(q.matrix, coord, <span class="at">method =</span> <span class="st">"map.max"</span>, <span class="at">interpol =</span> <span class="fu">FieldsKrigModel</span>(<span class="dv">10</span>),  </span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Ancestry coefficients"</span>,</span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Longitude"</span>, <span class="at">ylab =</span> <span class="st">"Latitude"</span>, </span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>     <span class="at">resolution =</span> <span class="fu">c</span>(<span class="dv">300</span>,<span class="dv">300</span>), <span class="at">cex =</span> .<span class="dv">4</span>, </span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>     <span class="at">col.palette =</span> pal)</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a>map<span class="ot">&lt;-</span><span class="fu">map</span>(<span class="at">database=</span> <span class="st">"county"</span>, <span class="at">add =</span> T, <span class="at">interior =</span> T, <span class="at">col=</span><span class="st">"grey"</span>,<span class="at">lwd=</span><span class="fl">0.2</span>)</span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a>map<span class="ot">&lt;-</span><span class="fu">map</span>(<span class="at">database=</span> <span class="st">"state"</span>, <span class="at">add =</span> T, <span class="at">interior =</span> T, <span class="at">col=</span><span class="st">"grey"</span>,<span class="at">lwd=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="geospatial-analysis" class="level1">
<h1>Geospatial analysis</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(terra)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SDMtune)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mapview)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Quercus_bicolor/geospatial/"</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co">#get bioclimatic data from https://www.worldclim.org/data/worldclim21.html</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co">#put in "~/Quercus_bicolor/geospatial/env_data/"</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="co">#assemble occurences and create buffers</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>qubi<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">"Qbicolor_occurences.csv"</span>)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>qubi<span class="ot">&lt;-</span>qubi[qubi<span class="sc">$</span>pa<span class="sc">==</span><span class="dv">1</span>,]</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>qubi<span class="ot">&lt;-</span><span class="fu">vect</span>(qubi, <span class="at">crs=</span><span class="st">'EPSG:4326'</span>, <span class="fu">c</span>(<span class="st">'x'</span>,<span class="st">'y'</span>))</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>qubi_buf500<span class="ot">&lt;-</span><span class="fu">buffer</span>(qubi, <span class="at">width=</span><span class="dv">500000</span>)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>qubi_buf500<span class="ot">&lt;-</span><span class="fu">aggregate</span>(qubi_buf500, <span class="at">dissolve=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>quly<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">"Qlyrata_occurences.csv"</span>)</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(quly)</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>quly<span class="ot">&lt;-</span><span class="fu">vect</span>(quly, <span class="at">crs=</span><span class="st">'EPSG:4326'</span>, <span class="fu">c</span>(<span class="st">'decimalLongitude'</span>,<span class="st">'decimalLatitude'</span>))</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>quly_buf500<span class="ot">&lt;-</span><span class="fu">buffer</span>(quly, <span class="at">width=</span><span class="dv">500000</span>)</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>quly_buf500<span class="ot">&lt;-</span><span class="fu">aggregate</span>(quly_buf500, <span class="at">dissolve=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(quly_buf500)</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>merge_buf500<span class="ot">&lt;-</span><span class="fu">rbind</span>(quly_buf500, qubi_buf500)</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>merge_buf500<span class="ot">&lt;-</span><span class="fu">aggregate</span>(merge_buf500, <span class="at">dissolve=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>qubi_buf150<span class="ot">&lt;-</span><span class="fu">buffer</span>(qubi, <span class="at">width=</span><span class="dv">150000</span>)</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>qubi_buf150<span class="ot">&lt;-</span><span class="fu">fillHoles</span>(qubi_buf150)</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(qubi_buf150)</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>quly_buf150<span class="ot">&lt;-</span><span class="fu">buffer</span>(quly, <span class="at">width=</span><span class="dv">150000</span>)</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>quly_buf150<span class="ot">&lt;-</span><span class="fu">aggregate</span>(quly_buf150, <span class="at">dissolve=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>quly_buf150<span class="ot">&lt;-</span><span class="fu">fillHoles</span>(quly_buf150)</span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(quly_buf150)</span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>merge_buf150<span class="ot">&lt;-</span><span class="fu">rbind</span>(quly_buf150, qubi_buf150)</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>merge_buf150<span class="ot">&lt;-</span><span class="fu">aggregate</span>(merge_buf150, <span class="at">dissolve=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(merge_buf150)</span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a>merge_buf150<span class="ot">&lt;-</span><span class="fu">fillHoles</span>(merge_buf150)</span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a><span class="fu">writeVector</span>(merge_buf150, <span class="st">"merge_buf150.shp"</span>)</span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a><span class="fu">writeVector</span>(qubi_buf150, <span class="st">"qubi_buf150.shp"</span>)</span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a><span class="fu">writeVector</span>(quly_buf150, <span class="st">"quly_buf150.shp"</span>)</span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a>output_folder <span class="ot">&lt;-</span><span class="st">"~Quercus_bicolor/geospatial/"</span></span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(output_folder, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.tif$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a>bioclim2<span class="ot">&lt;-</span><span class="fu">rast</span>(files)</span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a><span class="fu">crs</span>(bioclim)<span class="sc">==</span><span class="fu">crs</span>(bioclim2)</span>
<span id="cb25-41"><a href="#cb25-41" aria-hidden="true" tabindex="-1"></a>bioclim<span class="ot">&lt;-</span><span class="fu">rast</span>(files)</span>
<span id="cb25-42"><a href="#cb25-42" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bioclim2<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_1, )</span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a>ex<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">105</span>,<span class="sc">-</span><span class="dv">69</span>,<span class="dv">20</span>,<span class="dv">50</span>)</span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a>bioclimc<span class="ot">&lt;-</span><span class="fu">crop</span>(bioclim2,ex)</span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bioclimc<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_1)</span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a><span class="do">##model SWO</span></span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a>bioclim_qubi<span class="ot">&lt;-</span><span class="fu">crop</span>(bioclim, qubi_buf150, <span class="at">mask=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bioclim_qubi<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_1)</span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a>qubi_bg<span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">spatSample</span>(bioclim_qubi, <span class="at">size =</span> <span class="dv">10000</span>, <span class="at">method =</span> <span class="st">"random"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>,<span class="at">xy =</span> <span class="cn">TRUE</span>,<span class="at">values =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a>qubi_pres<span class="ot">&lt;-</span><span class="fu">geom</span>(qubi)[,<span class="fu">c</span>(<span class="st">'x'</span>,<span class="st">'y'</span>)] <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a>qubi_bg<span class="ot">&lt;-</span>qubi_bg <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb25-53"><a href="#cb25-53" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(zeallot)</span>
<span id="cb25-54"><a href="#cb25-54" aria-hidden="true" tabindex="-1"></a>qubi_pres<span class="ot">&lt;-</span><span class="fu">thinData</span>(qubi_pres, bioclim_qubi, <span class="at">x=</span> <span class="st">'x'</span>, <span class="at">y=</span><span class="st">'y'</span>)</span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a>qubi_bg<span class="ot">&lt;-</span><span class="fu">thinData</span>(qubi_bg, bioclim_qubi, <span class="at">x=</span> <span class="st">'x'</span>, <span class="at">y=</span><span class="st">'y'</span>)</span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a>bg<span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">spatSample</span>(bioclim_qubi, <span class="at">size =</span> <span class="dv">10000</span>, <span class="at">method =</span> <span class="st">"random"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>,<span class="at">xy =</span> <span class="cn">TRUE</span>,<span class="at">values =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a>bgSWD<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"background"</span>,bioclim_qubi, <span class="at">p=</span><span class="cn">NULL</span>,<span class="at">a=</span>bg)</span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a>data<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"quercusbicolor"</span>, bioclim_qubi, <span class="at">p=</span>qubi_pres, <span class="at">a=</span>qubi_bg )</span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(qubi_train, qubi_test) <span class="sc">%&lt;-%</span> <span class="fu">trainValTest</span>(data, </span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">test =</span> <span class="fl">0.15</span>, </span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">only_presence =</span> <span class="cn">TRUE</span>, </span>
<span id="cb25-62"><a href="#cb25-62" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">seed =</span> <span class="dv">33</span>)</span>
<span id="cb25-63"><a href="#cb25-63" aria-hidden="true" tabindex="-1"></a><span class="co">#parition training data</span></span>
<span id="cb25-64"><a href="#cb25-64" aria-hidden="true" tabindex="-1"></a>cv<span class="ot">&lt;-</span><span class="fu">randomFolds</span>(qubi_train, <span class="at">k=</span><span class="dv">10</span>, <span class="at">only_presence=</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-65"><a href="#cb25-65" aria-hidden="true" tabindex="-1"></a><span class="co">#create maxent model with default values</span></span>
<span id="cb25-66"><a href="#cb25-66" aria-hidden="true" tabindex="-1"></a>maxent_all<span class="ot">&lt;-</span><span class="fu">train</span>(<span class="st">"Maxent"</span>, qubi_train, <span class="at">folds=</span>cv, <span class="at">progress=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-67"><a href="#cb25-67" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training TSS: "</span>, <span class="fu">auc</span>(maxent_all))</span>
<span id="cb25-68"><a href="#cb25-68" aria-hidden="true" tabindex="-1"></a>selected_variables_model <span class="ot">&lt;-</span> <span class="fu">varSel</span>(maxent_all, <span class="at">metric =</span> <span class="st">"auc"</span>,<span class="at">test=</span>qubi_test,<span class="at">bg4cor =</span> bgSWD, <span class="at">method =</span> <span class="st">"spearman"</span>, <span class="at">cor_th =</span> <span class="fl">0.7</span>, <span class="at">permut =</span> <span class="dv">1</span>,<span class="at">progress=</span><span class="cn">TRUE</span>,<span class="at">use_pc=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-69"><a href="#cb25-69" aria-hidden="true" tabindex="-1"></a>va<span class="ot">&lt;-</span><span class="fu">varImp</span>(selected_variables_model)</span>
<span id="cb25-70"><a href="#cb25-70" aria-hidden="true" tabindex="-1"></a><span class="fu">plotVarImp</span>(va)</span>
<span id="cb25-71"><a href="#cb25-71" aria-hidden="true" tabindex="-1"></a><span class="co">#reduced_variables_model</span></span>
<span id="cb25-72"><a href="#cb25-72" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training AUC: "</span>, <span class="fu">auc</span>(reduced_variables_model))</span>
<span id="cb25-73"><a href="#cb25-73" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing AUC: "</span>, <span class="fu">auc</span>(reduced_variables_model, <span class="at">test =</span> qubi_test))</span>
<span id="cb25-74"><a href="#cb25-74" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training TSS: "</span>, <span class="fu">tss</span>(reduced_variables_model))</span>
<span id="cb25-75"><a href="#cb25-75" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing TSS: "</span>, <span class="fu">tss</span>(reduced_variables_model, <span class="at">test =</span> qubi_test))</span>
<span id="cb25-76"><a href="#cb25-76" aria-hidden="true" tabindex="-1"></a>h_max <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">reg =</span> <span class="fu">seq</span>(<span class="fl">0.2</span>,<span class="dv">2</span>,<span class="fl">0.2</span>), <span class="at">fc =</span> <span class="fu">c</span>(<span class="st">"l"</span>, <span class="st">"lq"</span>, <span class="st">"lh"</span>, <span class="st">"lqp"</span>, <span class="st">"lqph"</span>, <span class="st">"lqpht"</span>), <span class="at">iter=</span> <span class="fu">seq</span>(<span class="dv">300</span>,<span class="dv">1500</span>,<span class="dv">200</span>))</span>
<span id="cb25-77"><a href="#cb25-77" aria-hidden="true" tabindex="-1"></a><span class="co"># Test all the possible combinations with gridSearch</span></span>
<span id="cb25-78"><a href="#cb25-78" aria-hidden="true" tabindex="-1"></a>gs_maxall<span class="ot">&lt;-</span> <span class="fu">gridSearch</span>(selected_variables_model, <span class="at">hypers =</span> h_max, <span class="at">metric =</span> <span class="st">"tss"</span>, <span class="at">save_models=</span><span class="cn">FALSE</span>, <span class="at">interactive=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-79"><a href="#cb25-79" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gs_maxall<span class="sc">@</span>results[<span class="fu">order</span>(<span class="sc">-</span>gs_maxall<span class="sc">@</span>results<span class="sc">$</span>train_TSS), ])  <span class="co"># Best combinations</span></span>
<span id="cb25-80"><a href="#cb25-80" aria-hidden="true" tabindex="-1"></a>gs_maxall<span class="sc">@</span>results[<span class="fu">order</span>(<span class="sc">-</span>gs_maxall<span class="sc">@</span>results<span class="sc">$</span>test_TSS), ]  <span class="co"># Best combinations</span></span>
<span id="cb25-81"><a href="#cb25-81" aria-hidden="true" tabindex="-1"></a>gs_maxall<span class="sc">@</span>results<span class="sc">$</span>sum<span class="ot">&lt;-</span>gs_maxall<span class="sc">@</span>results<span class="sc">$</span>train_TSS<span class="sc">+</span>gs_maxall<span class="sc">@</span>results<span class="sc">$</span>test_TSS</span>
<span id="cb25-82"><a href="#cb25-82" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gs_maxall<span class="sc">@</span>results[<span class="fu">order</span>(<span class="sc">-</span>gs_maxall<span class="sc">@</span>results<span class="sc">$</span>sum), ])  <span class="co"># Best combinations</span></span>
<span id="cb25-83"><a href="#cb25-83" aria-hidden="true" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="fu">which.max</span>(gs_max<span class="sc">@</span>results<span class="sc">$</span>test_TSS)</span>
<span id="cb25-84"><a href="#cb25-84" aria-hidden="true" tabindex="-1"></a>vars<span class="ot">&lt;-</span><span class="fu">c</span>(bioclim_qubi<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_2 ,bioclim_qubi<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_6 ,bioclim_qubi<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_8 ,bioclim_qubi<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_5 ,bioclim_qubi<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_15 ,bioclim_qubi<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_18)</span>
<span id="cb25-85"><a href="#cb25-85" aria-hidden="true" tabindex="-1"></a>data<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"quercusbicolor"</span>, vars, <span class="at">p=</span>qubi_pres, <span class="at">a=</span>qubi_bg)</span>
<span id="cb25-86"><a href="#cb25-86" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(qubi_train, qubi_test) <span class="sc">%&lt;-%</span> <span class="fu">trainValTest</span>(data, </span>
<span id="cb25-87"><a href="#cb25-87" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">test =</span> <span class="fl">0.2</span>, </span>
<span id="cb25-88"><a href="#cb25-88" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">only_presence =</span> <span class="cn">TRUE</span>, </span>
<span id="cb25-89"><a href="#cb25-89" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">seed =</span> <span class="dv">39</span>)</span>
<span id="cb25-90"><a href="#cb25-90" aria-hidden="true" tabindex="-1"></a><span class="co">#parition training data</span></span>
<span id="cb25-91"><a href="#cb25-91" aria-hidden="true" tabindex="-1"></a>cv<span class="ot">&lt;-</span><span class="fu">randomFolds</span>(qubi_train, <span class="at">k=</span><span class="dv">10</span>, <span class="at">only_presence=</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-92"><a href="#cb25-92" aria-hidden="true" tabindex="-1"></a><span class="co">#select optimal parameters</span></span>
<span id="cb25-93"><a href="#cb25-93" aria-hidden="true" tabindex="-1"></a>max_finalcv<span class="ot">&lt;-</span><span class="fu">train</span>(<span class="st">"Maxent"</span>, </span>
<span id="cb25-94"><a href="#cb25-94" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> qubi_train,</span>
<span id="cb25-95"><a href="#cb25-95" aria-hidden="true" tabindex="-1"></a>                   <span class="at">folds=</span>cv,</span>
<span id="cb25-96"><a href="#cb25-96" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fc =</span> <span class="st">"lqpht"</span>, </span>
<span id="cb25-97"><a href="#cb25-97" aria-hidden="true" tabindex="-1"></a>                   <span class="at">reg =</span> <span class="fl">0.6</span>,</span>
<span id="cb25-98"><a href="#cb25-98" aria-hidden="true" tabindex="-1"></a>                   <span class="at">iter =</span> <span class="dv">900</span>)</span>
<span id="cb25-99"><a href="#cb25-99" aria-hidden="true" tabindex="-1"></a>max_final_qubi<span class="ot">&lt;-</span><span class="fu">combineCV</span>(max_finalcv)</span>
<span id="cb25-100"><a href="#cb25-100" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training AUC: "</span>, <span class="fu">auc</span>(max_final_qubi))</span>
<span id="cb25-101"><a href="#cb25-101" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing AUC: "</span>, <span class="fu">auc</span>(max_final_qubi, <span class="at">test =</span> qubi_test))</span>
<span id="cb25-102"><a href="#cb25-102" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training TSS: "</span>, <span class="fu">tss</span>(max_final_qubi))</span>
<span id="cb25-103"><a href="#cb25-103" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing TSS: "</span>, <span class="fu">tss</span>(max_final_qubi, <span class="at">test =</span> qubi_test))</span>
<span id="cb25-104"><a href="#cb25-104" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"D:/data/wc2.1_30s_bio/raw_rasters_touse/QUBI_and_QULY/modeling_outputs"</span>)</span>
<span id="cb25-105"><a href="#cb25-105" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span>
<span id="cb25-106"><a href="#cb25-106" aria-hidden="true" tabindex="-1"></a>final_max_modelreport<span class="ot">&lt;-</span><span class="fu">modelReport</span>(max_final_qubi, <span class="at">folder=</span><span class="st">"maxmodelreport"</span>, <span class="at">test =</span> qubi_test,<span class="at">type =</span> <span class="st">"cloglog"</span>,<span class="at">response_curves =</span> <span class="cn">TRUE</span>,<span class="at">only_presence =</span> <span class="cn">TRUE</span>,<span class="at">jk =</span> <span class="cn">TRUE</span>,<span class="at">env =</span> vars, <span class="at">clamp =</span> <span class="cn">TRUE</span>, <span class="at">permut =</span> <span class="dv">10</span>,<span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-107"><a href="#cb25-107" aria-hidden="true" tabindex="-1"></a><span class="co">#get paleo data</span></span>
<span id="cb25-108"><a href="#cb25-108" aria-hidden="true" tabindex="-1"></a>output_folder <span class="ot">&lt;-</span><span class="st">"D:/data/paleoclimate/"</span></span>
<span id="cb25-109"><a href="#cb25-109" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(output_folder, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.tif$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-110"><a href="#cb25-110" aria-hidden="true" tabindex="-1"></a>nontemp<span class="ot">&lt;-</span><span class="fu">rast</span>(files)</span>
<span id="cb25-111"><a href="#cb25-111" aria-hidden="true" tabindex="-1"></a>pal<span class="ot">&lt;-</span>temp<span class="sc">/</span><span class="dv">10</span></span>
<span id="cb25-112"><a href="#cb25-112" aria-hidden="true" tabindex="-1"></a>paleo<span class="ot">&lt;-</span><span class="fu">c</span>(pal,nontemp)</span>
<span id="cb25-113"><a href="#cb25-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-114"><a href="#cb25-114" aria-hidden="true" tabindex="-1"></a>paleo<span class="ot">&lt;-</span><span class="fu">project</span>(paleo, bioclimc, <span class="at">method=</span> <span class="st">'bilinear'</span>, <span class="at">mask=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-115"><a href="#cb25-115" aria-hidden="true" tabindex="-1"></a><span class="fu">crs</span>(paleo)<span class="sc">==</span><span class="fu">crs</span>(bioclimc) <span class="co">#if no, will need to re project</span></span>
<span id="cb25-116"><a href="#cb25-116" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(paleo)<span class="sc">==</span><span class="fu">ext</span>(bioclimc)</span>
<span id="cb25-117"><a href="#cb25-117" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(paleo<span class="sc">$</span>cclgmbi1)</span>
<span id="cb25-118"><a href="#cb25-118" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(paleo)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"wc2.1_2.5m_bio_1"</span>, <span class="st">"wc2.1_2.5m_bio_10"</span>, <span class="st">"wc2.1_2.5m_bio_11"</span>,<span class="st">"wc2.1_2.5m_bio_2"</span> ,<span class="st">"wc2.1_2.5m_bio_3"</span>,<span class="st">"wc2.1_2.5m_bio_4"</span>,<span class="st">"wc2.1_2.5m_bio_5"</span>,<span class="st">"wc2.1_2.5m_bio_6"</span> ,<span class="st">"wc2.1_2.5m_bio_7"</span>,<span class="st">"wc2.1_2.5m_bio_8"</span>,<span class="st">"wc2.1_2.5m_bio_9"</span>,<span class="st">"wc2.1_2.5m_bio_12"</span>,<span class="st">"wc2.1_2.5m_bio_13"</span>, <span class="st">"wc2.1_2.5m_bio_14"</span>, <span class="st">"wc2.1_2.5m_bio_15"</span>, <span class="st">"wc2.1_2.5m_bio_16"</span>, <span class="st">"wc2.1_2.5m_bio_17"</span>, <span class="st">"wc2.1_2.5m_bio_18"</span>, <span class="st">"wc2.1_2.5m_bio_19"</span>)</span>
<span id="cb25-119"><a href="#cb25-119" aria-hidden="true" tabindex="-1"></a><span class="fu">writeRaster</span>(paleo,<span class="st">"LGMclimate_temp_converted.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-120"><a href="#cb25-120" aria-hidden="true" tabindex="-1"></a>max_predict_qubi_current<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_qubi, <span class="at">data=</span>bioclimc, <span class="at">filename =</span> <span class="st">"max_predict_qubi_current_fullrange.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-121"><a href="#cb25-121" aria-hidden="true" tabindex="-1"></a>max_predict_qubi_LGM<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_qubi, <span class="at">data=</span>paleo, <span class="at">filename =</span> <span class="st">"max_predict_qubi_LGM.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-122"><a href="#cb25-122" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_current)</span>
<span id="cb25-123"><a href="#cb25-123" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_LGM)</span>
<span id="cb25-124"><a href="#cb25-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-125"><a href="#cb25-125" aria-hidden="true" tabindex="-1"></a><span class="co">#get interglacial data</span></span>
<span id="cb25-126"><a href="#cb25-126" aria-hidden="true" tabindex="-1"></a>output_folder <span class="ot">&lt;-</span><span class="st">"D:/data/interglacialclimate/temp"</span></span>
<span id="cb25-127"><a href="#cb25-127" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(output_folder)</span>
<span id="cb25-128"><a href="#cb25-128" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(output_folder, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.bil$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-129"><a href="#cb25-129" aria-hidden="true" tabindex="-1"></a>nontemp<span class="ot">&lt;-</span><span class="fu">rast</span>(files)</span>
<span id="cb25-130"><a href="#cb25-130" aria-hidden="true" tabindex="-1"></a>int<span class="ot">&lt;-</span>interglacial<span class="sc">/</span><span class="dv">10</span></span>
<span id="cb25-131"><a href="#cb25-131" aria-hidden="true" tabindex="-1"></a>interglacial<span class="ot">&lt;-</span><span class="fu">c</span>(int, nontemp)</span>
<span id="cb25-132"><a href="#cb25-132" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(interglacial<span class="sc">$</span>lig_30s_bio_1)</span>
<span id="cb25-133"><a href="#cb25-133" aria-hidden="true" tabindex="-1"></a><span class="fu">crs</span>(interglacial)<span class="sc">==</span><span class="fu">crs</span>(bioclimc) <span class="co">#is false proceed with next 2 lines</span></span>
<span id="cb25-134"><a href="#cb25-134" aria-hidden="true" tabindex="-1"></a>interglacial<span class="ot">&lt;-</span><span class="fu">project</span>(interglacial, bioclimc, <span class="at">method=</span> <span class="st">'bilinear'</span>, <span class="at">mask=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-135"><a href="#cb25-135" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(interglacial)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"wc2.1_2.5m_bio_1"</span>, <span class="st">"wc2.1_2.5m_bio_10"</span>, <span class="st">"wc2.1_2.5m_bio_11"</span>,<span class="st">"wc2.1_2.5m_bio_2"</span> ,<span class="st">"wc2.1_2.5m_bio_3"</span>,<span class="st">"wc2.1_2.5m_bio_4"</span>,<span class="st">"wc2.1_2.5m_bio_5"</span>,<span class="st">"wc2.1_2.5m_bio_6"</span> ,<span class="st">"wc2.1_2.5m_bio_7"</span>,<span class="st">"wc2.1_2.5m_bio_8"</span>,<span class="st">"wc2.1_2.5m_bio_9"</span>,<span class="st">"wc2.1_2.5m_bio_12"</span>,<span class="st">"wc2.1_2.5m_bio_13"</span>, <span class="st">"wc2.1_2.5m_bio_14"</span>, <span class="st">"wc2.1_2.5m_bio_15"</span>, <span class="st">"wc2.1_2.5m_bio_16"</span>, <span class="st">"wc2.1_2.5m_bio_17"</span>, <span class="st">"wc2.1_2.5m_bio_18"</span>, <span class="st">"wc2.1_2.5m_bio_19"</span>)</span>
<span id="cb25-136"><a href="#cb25-136" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(interglacial<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_1)</span>
<span id="cb25-137"><a href="#cb25-137" aria-hidden="true" tabindex="-1"></a><span class="fu">writeRaster</span>(interglacial,<span class="st">"interglacial_converted.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-138"><a href="#cb25-138" aria-hidden="true" tabindex="-1"></a><span class="co">#predict</span></span>
<span id="cb25-139"><a href="#cb25-139" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>()</span>
<span id="cb25-140"><a href="#cb25-140" aria-hidden="true" tabindex="-1"></a>max_predict_qubi_IG<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_qubi, <span class="at">data=</span>interglacial, <span class="at">filename =</span> <span class="st">"max_predict_qubi_IG.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-141"><a href="#cb25-141" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_IG)</span>
<span id="cb25-142"><a href="#cb25-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-143"><a href="#cb25-143" aria-hidden="true" tabindex="-1"></a><span class="co">#get midholocene data</span></span>
<span id="cb25-144"><a href="#cb25-144" aria-hidden="true" tabindex="-1"></a>output_folder <span class="ot">&lt;-</span><span class="st">"D:/data/midholocene/"</span></span>
<span id="cb25-145"><a href="#cb25-145" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(output_folder)</span>
<span id="cb25-146"><a href="#cb25-146" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(output_folder, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.tif$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-147"><a href="#cb25-147" aria-hidden="true" tabindex="-1"></a>nontemp<span class="ot">&lt;-</span><span class="fu">rast</span>(files)</span>
<span id="cb25-148"><a href="#cb25-148" aria-hidden="true" tabindex="-1"></a>int<span class="ot">&lt;-</span>temp<span class="sc">/</span><span class="dv">10</span></span>
<span id="cb25-149"><a href="#cb25-149" aria-hidden="true" tabindex="-1"></a>midhol<span class="ot">&lt;-</span><span class="fu">c</span>(int, nontemp)</span>
<span id="cb25-150"><a href="#cb25-150" aria-hidden="true" tabindex="-1"></a><span class="fu">crs</span>(midhol)<span class="sc">==</span><span class="fu">crs</span>(bioclimc) <span class="co">#is false proceed with next 2 lines</span></span>
<span id="cb25-151"><a href="#cb25-151" aria-hidden="true" tabindex="-1"></a>midhol<span class="ot">&lt;-</span><span class="fu">project</span>(midhol, bioclimc, <span class="at">method=</span> <span class="st">'bilinear'</span>, <span class="at">mask=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-152"><a href="#cb25-152" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(midhol)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"wc2.1_2.5m_bio_1"</span>, <span class="st">"wc2.1_2.5m_bio_10"</span>, <span class="st">"wc2.1_2.5m_bio_11"</span>,<span class="st">"wc2.1_2.5m_bio_2"</span> ,<span class="st">"wc2.1_2.5m_bio_3"</span>,<span class="st">"wc2.1_2.5m_bio_4"</span>,<span class="st">"wc2.1_2.5m_bio_5"</span>,<span class="st">"wc2.1_2.5m_bio_6"</span> ,<span class="st">"wc2.1_2.5m_bio_7"</span>,<span class="st">"wc2.1_2.5m_bio_8"</span>,<span class="st">"wc2.1_2.5m_bio_9"</span>,<span class="st">"wc2.1_2.5m_bio_12"</span>,<span class="st">"wc2.1_2.5m_bio_13"</span>, <span class="st">"wc2.1_2.5m_bio_14"</span>, <span class="st">"wc2.1_2.5m_bio_15"</span>, <span class="st">"wc2.1_2.5m_bio_16"</span>, <span class="st">"wc2.1_2.5m_bio_17"</span>, <span class="st">"wc2.1_2.5m_bio_18"</span>, <span class="st">"wc2.1_2.5m_bio_19"</span>)</span>
<span id="cb25-153"><a href="#cb25-153" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(midhol<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_1)</span>
<span id="cb25-154"><a href="#cb25-154" aria-hidden="true" tabindex="-1"></a><span class="fu">writeRaster</span>(midhol,<span class="st">"midhol_converted.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-155"><a href="#cb25-155" aria-hidden="true" tabindex="-1"></a><span class="co">#predict</span></span>
<span id="cb25-156"><a href="#cb25-156" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>()</span>
<span id="cb25-157"><a href="#cb25-157" aria-hidden="true" tabindex="-1"></a>max_predict_qubi_midhol<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_qubi, <span class="at">data=</span>midhol, <span class="at">filename =</span> <span class="st">"max_predict_qubi_midholocene.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-158"><a href="#cb25-158" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_midhol)</span>
<span id="cb25-159"><a href="#cb25-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-160"><a href="#cb25-160" aria-hidden="true" tabindex="-1"></a><span class="do">##model OVERCUP OAK</span></span>
<span id="cb25-161"><a href="#cb25-161" aria-hidden="true" tabindex="-1"></a>bioclim_quly<span class="ot">&lt;-</span><span class="fu">crop</span>(bioclim, quly_buf150, <span class="at">mask=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-162"><a href="#cb25-162" aria-hidden="true" tabindex="-1"></a>quly_bg<span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">spatSample</span>(bioclim_quly, <span class="at">size =</span> <span class="dv">10000</span>, <span class="at">method =</span> <span class="st">"random"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>,<span class="at">xy =</span> <span class="cn">TRUE</span>,<span class="at">values =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-163"><a href="#cb25-163" aria-hidden="true" tabindex="-1"></a>quly_pres<span class="ot">&lt;-</span><span class="fu">geom</span>(quly)[,<span class="fu">c</span>(<span class="st">'x'</span>,<span class="st">'y'</span>)] <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb25-164"><a href="#cb25-164" aria-hidden="true" tabindex="-1"></a>quly_bg<span class="ot">&lt;-</span>quly_bg <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb25-165"><a href="#cb25-165" aria-hidden="true" tabindex="-1"></a>quly_pres<span class="ot">&lt;-</span><span class="fu">thinData</span>(quly_pres, bioclim_quly, <span class="at">x=</span> <span class="st">'x'</span>, <span class="at">y=</span><span class="st">'y'</span>)</span>
<span id="cb25-166"><a href="#cb25-166" aria-hidden="true" tabindex="-1"></a>quly_bg<span class="ot">&lt;-</span><span class="fu">thinData</span>(quly_bg, bioclim_quly, <span class="at">x=</span> <span class="st">'x'</span>, <span class="at">y=</span><span class="st">'y'</span>)</span>
<span id="cb25-167"><a href="#cb25-167" aria-hidden="true" tabindex="-1"></a>bg<span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">spatSample</span>(bioclim_quly, <span class="at">size =</span> <span class="dv">10000</span>, <span class="at">method =</span> <span class="st">"random"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>,<span class="at">xy =</span> <span class="cn">TRUE</span>,<span class="at">values =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-168"><a href="#cb25-168" aria-hidden="true" tabindex="-1"></a>bgSWD<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"background"</span>,bioclim_quly, <span class="at">p=</span><span class="cn">NULL</span>,<span class="at">a=</span>bg)</span>
<span id="cb25-169"><a href="#cb25-169" aria-hidden="true" tabindex="-1"></a>data<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"quercuslyrata"</span>, bioclim_quly, <span class="at">p=</span>quly_pres, <span class="at">a=</span>quly_bg )</span>
<span id="cb25-170"><a href="#cb25-170" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(quly_train, quly_test) <span class="sc">%&lt;-%</span> <span class="fu">trainValTest</span>(data, </span>
<span id="cb25-171"><a href="#cb25-171" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">test =</span> <span class="fl">0.15</span>, </span>
<span id="cb25-172"><a href="#cb25-172" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">only_presence =</span> <span class="cn">TRUE</span>, </span>
<span id="cb25-173"><a href="#cb25-173" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">seed =</span> <span class="dv">33</span>)</span>
<span id="cb25-174"><a href="#cb25-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-175"><a href="#cb25-175" aria-hidden="true" tabindex="-1"></a><span class="co">#parition training data</span></span>
<span id="cb25-176"><a href="#cb25-176" aria-hidden="true" tabindex="-1"></a>cv<span class="ot">&lt;-</span><span class="fu">randomFolds</span>(quly_train, <span class="at">k=</span><span class="dv">10</span>, <span class="at">only_presence=</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-177"><a href="#cb25-177" aria-hidden="true" tabindex="-1"></a><span class="co">#create maxent model with default values</span></span>
<span id="cb25-178"><a href="#cb25-178" aria-hidden="true" tabindex="-1"></a>maxent_quly<span class="ot">&lt;-</span><span class="fu">train</span>(<span class="st">"Maxent"</span>, quly_train, <span class="at">folds=</span>cv, <span class="at">progress=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-179"><a href="#cb25-179" aria-hidden="true" tabindex="-1"></a>selected_variables_model_quly <span class="ot">&lt;-</span> <span class="fu">varSel</span>(maxent_quly, <span class="at">metric =</span> <span class="st">"auc"</span>,<span class="at">test=</span>quly_test,<span class="at">bg4cor =</span> bgSWD, <span class="at">method =</span> <span class="st">"spearman"</span>, <span class="at">cor_th =</span> <span class="fl">0.7</span>, <span class="at">permut =</span> <span class="dv">1</span>,<span class="at">progress=</span><span class="cn">TRUE</span>,<span class="at">use_pc=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-180"><a href="#cb25-180" aria-hidden="true" tabindex="-1"></a>vars_quly<span class="ot">&lt;-</span><span class="fu">c</span>(bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_10 ,bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_13 ,bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_15 ,bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_17 ,bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_18 ,bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_2, bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_3, bioclim_quly<span class="sc">$</span>wc2<span class="fl">.1</span>_2<span class="fl">.5</span>m_bio_8)</span>
<span id="cb25-181"><a href="#cb25-181" aria-hidden="true" tabindex="-1"></a><span class="co">#reduced_variables_model</span></span>
<span id="cb25-182"><a href="#cb25-182" aria-hidden="true" tabindex="-1"></a>h_max <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">reg =</span> <span class="fu">seq</span>(<span class="fl">0.2</span>,<span class="dv">2</span>,<span class="fl">0.2</span>), <span class="at">fc =</span> <span class="fu">c</span>(<span class="st">"l"</span>, <span class="st">"lq"</span>, <span class="st">"lh"</span>, <span class="st">"lqp"</span>, <span class="st">"lqph"</span>, <span class="st">"lqpht"</span>), <span class="at">iter=</span> <span class="fu">seq</span>(<span class="dv">300</span>,<span class="dv">1500</span>,<span class="dv">200</span>))</span>
<span id="cb25-183"><a href="#cb25-183" aria-hidden="true" tabindex="-1"></a><span class="co"># Test all the possible combinations with gridSearch</span></span>
<span id="cb25-184"><a href="#cb25-184" aria-hidden="true" tabindex="-1"></a>gs_max_quly<span class="ot">&lt;-</span> <span class="fu">gridSearch</span>(selected_variables_model, <span class="at">hypers =</span> h_max, <span class="at">metric =</span> <span class="st">"tss"</span>, <span class="at">save_models=</span><span class="cn">FALSE</span>, <span class="at">interactive=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-185"><a href="#cb25-185" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gs_max_quly<span class="sc">@</span>results[<span class="fu">order</span>(gs_max_quly<span class="sc">@</span>results<span class="sc">$</span>diff_TSS), ])  <span class="co"># Best combinations</span></span>
<span id="cb25-186"><a href="#cb25-186" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gs_max_quly<span class="sc">@</span>results[<span class="fu">order</span>(<span class="sc">-</span>gs_max_quly<span class="sc">@</span>results<span class="sc">$</span>train_TSS), ])  <span class="co"># Best combinations</span></span>
<span id="cb25-187"><a href="#cb25-187" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gs_max_quly<span class="sc">@</span>results[<span class="fu">order</span>(<span class="sc">-</span>gs_max_quly<span class="sc">@</span>results<span class="sc">$</span>test_TSS), ])  <span class="co"># Best combinations</span></span>
<span id="cb25-188"><a href="#cb25-188" aria-hidden="true" tabindex="-1"></a>gs_max_quly<span class="sc">@</span>results<span class="sc">$</span>sum<span class="ot">&lt;-</span>gs_max_quly<span class="sc">@</span>results<span class="sc">$</span>train_TSS<span class="sc">+</span>gs_max_quly<span class="sc">@</span>results<span class="sc">$</span>test_TSS</span>
<span id="cb25-189"><a href="#cb25-189" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gs_max_quly<span class="sc">@</span>results[<span class="fu">order</span>(<span class="sc">-</span>gs_max_quly<span class="sc">@</span>results<span class="sc">$</span>sum), ])  <span class="co"># Best combinations</span></span>
<span id="cb25-190"><a href="#cb25-190" aria-hidden="true" tabindex="-1"></a><span class="co">#select optimal parameters</span></span>
<span id="cb25-191"><a href="#cb25-191" aria-hidden="true" tabindex="-1"></a>data<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"quercuslyrata"</span>, vars_quly, <span class="at">p=</span>quly_pres, <span class="at">a=</span>quly_bg )</span>
<span id="cb25-192"><a href="#cb25-192" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(quly_train, quly_test) <span class="sc">%&lt;-%</span> <span class="fu">trainValTest</span>(data, </span>
<span id="cb25-193"><a href="#cb25-193" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">test =</span> <span class="fl">0.2</span>, </span>
<span id="cb25-194"><a href="#cb25-194" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">only_presence =</span> <span class="cn">TRUE</span>, </span>
<span id="cb25-195"><a href="#cb25-195" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">seed =</span> <span class="dv">39</span>)</span>
<span id="cb25-196"><a href="#cb25-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-197"><a href="#cb25-197" aria-hidden="true" tabindex="-1"></a><span class="co">#parition training data</span></span>
<span id="cb25-198"><a href="#cb25-198" aria-hidden="true" tabindex="-1"></a>cv<span class="ot">&lt;-</span><span class="fu">randomFolds</span>(quly_train, <span class="at">k=</span><span class="dv">10</span>, <span class="at">only_presence=</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-199"><a href="#cb25-199" aria-hidden="true" tabindex="-1"></a>max_finalcv_quly<span class="ot">&lt;-</span><span class="fu">train</span>(<span class="st">"Maxent"</span>, </span>
<span id="cb25-200"><a href="#cb25-200" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> quly_train,</span>
<span id="cb25-201"><a href="#cb25-201" aria-hidden="true" tabindex="-1"></a>                   <span class="at">folds=</span>cv,</span>
<span id="cb25-202"><a href="#cb25-202" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fc =</span> <span class="st">"lqpht"</span>, </span>
<span id="cb25-203"><a href="#cb25-203" aria-hidden="true" tabindex="-1"></a>                   <span class="at">reg =</span> <span class="fl">0.6</span>,</span>
<span id="cb25-204"><a href="#cb25-204" aria-hidden="true" tabindex="-1"></a>                   <span class="at">iter =</span> <span class="dv">900</span>)</span>
<span id="cb25-205"><a href="#cb25-205" aria-hidden="true" tabindex="-1"></a>max_final_quly<span class="ot">&lt;-</span><span class="fu">combineCV</span>(max_finalcv_quly)</span>
<span id="cb25-206"><a href="#cb25-206" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training AUC: "</span>, <span class="fu">auc</span>(max_final_quly))</span>
<span id="cb25-207"><a href="#cb25-207" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing AUC: "</span>, <span class="fu">auc</span>(max_final_quly, <span class="at">test =</span> quly_test))</span>
<span id="cb25-208"><a href="#cb25-208" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training TSS: "</span>, <span class="fu">tss</span>(max_final_quly))</span>
<span id="cb25-209"><a href="#cb25-209" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing TSS: "</span>, <span class="fu">tss</span>(max_final_quly, <span class="at">test =</span> quly_test))</span>
<span id="cb25-210"><a href="#cb25-210" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"D:/data/wc2.1_30s_bio/raw_rasters_touse/QUBI_and_QULY/modeling_outputs"</span>)</span>
<span id="cb25-211"><a href="#cb25-211" aria-hidden="true" tabindex="-1"></a>final_max_modelreport<span class="ot">&lt;-</span><span class="fu">modelReport</span>(max_final_quly, <span class="at">folder=</span><span class="st">"max_final_quly_modreport"</span>, <span class="at">test =</span> quly_test,<span class="at">type =</span> <span class="st">"cloglog"</span>,<span class="at">response_curves =</span> <span class="cn">TRUE</span>,<span class="at">only_presence =</span> <span class="cn">TRUE</span>,<span class="at">jk =</span> <span class="cn">TRUE</span>,<span class="at">env =</span> vars_quly, <span class="at">clamp =</span> <span class="cn">TRUE</span>, <span class="at">permut =</span> <span class="dv">10</span>,<span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-212"><a href="#cb25-212" aria-hidden="true" tabindex="-1"></a><span class="co">#get paleo data for overcup</span></span>
<span id="cb25-213"><a href="#cb25-213" aria-hidden="true" tabindex="-1"></a>max_predict_quly_current<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_quly, <span class="at">data=</span>bioclimc, <span class="at">filename =</span> <span class="st">"max_predict_quly_current.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-214"><a href="#cb25-214" aria-hidden="true" tabindex="-1"></a>max_predict_quly_LGM<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_quly, <span class="at">data=</span>paleo, <span class="at">filename =</span> <span class="st">"max_predict_quly_LGM.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-215"><a href="#cb25-215" aria-hidden="true" tabindex="-1"></a>max_predict_quly_IG<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_quly, <span class="at">data=</span>interglacial, <span class="at">filename =</span> <span class="st">"max_predict_quly_IG.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-216"><a href="#cb25-216" aria-hidden="true" tabindex="-1"></a>max_predict_quly_midhol<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_quly, <span class="at">data=</span>midhol, <span class="at">filename =</span> <span class="st">"max_predict_quly_midholocene.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-217"><a href="#cb25-217" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_quly_midhol)</span>
<span id="cb25-218"><a href="#cb25-218" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_midhol)</span>
<span id="cb25-219"><a href="#cb25-219" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_quly_LGM)</span>
<span id="cb25-220"><a href="#cb25-220" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_LGM)</span>
<span id="cb25-221"><a href="#cb25-221" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_quly_IG)</span>
<span id="cb25-222"><a href="#cb25-222" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_IG)</span>
<span id="cb25-223"><a href="#cb25-223" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_current)</span>
<span id="cb25-224"><a href="#cb25-224" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_quly_current)</span>
<span id="cb25-225"><a href="#cb25-225" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geodata)</span>
<span id="cb25-226"><a href="#cb25-226" aria-hidden="true" tabindex="-1"></a>us <span class="ot">&lt;-</span> <span class="fu">gadm</span>(<span class="at">country =</span> <span class="st">"USA"</span>, <span class="at">level =</span> <span class="dv">2</span>, <span class="at">resolution =</span> <span class="dv">2</span>, <span class="at">path=</span><span class="fu">getwd</span>())</span>
<span id="cb25-227"><a href="#cb25-227" aria-hidden="true" tabindex="-1"></a>ten<span class="ot">&lt;-</span>us[us<span class="sc">$</span>NAME_1<span class="sc">==</span><span class="st">"Tennessee"</span>]</span>
<span id="cb25-228"><a href="#cb25-228" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ten, <span class="at">add=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-229"><a href="#cb25-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-230"><a href="#cb25-230" aria-hidden="true" tabindex="-1"></a><span class="co">#get future climate</span></span>
<span id="cb25-231"><a href="#cb25-231" aria-hidden="true" tabindex="-1"></a>fut<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"D:/data/future_climate/wc2.1_30s_bioc_EC-Earth3-Veg_ssp585_2041-2060.tif"</span>)</span>
<span id="cb25-232"><a href="#cb25-232" aria-hidden="true" tabindex="-1"></a><span class="fu">crs</span>(fut)<span class="sc">==</span><span class="fu">crs</span>(bioclimc)</span>
<span id="cb25-233"><a href="#cb25-233" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(fut)<span class="ot">=</span><span class="fu">c</span>(<span class="st">"wc2.1_2.5m_bio_1"</span>, <span class="st">"wc2.1_2.5m_bio_2"</span>,  <span class="st">"wc2.1_2.5m_bio_3"</span>,  <span class="st">"wc2.1_2.5m_bio_4"</span>,  <span class="st">"wc2.1_2.5m_bio_5"</span>, <span class="st">"wc2.1_2.5m_bio_6"</span>,  <span class="st">"wc2.1_2.5m_bio_7"</span>,  <span class="st">"wc2.1_2.5m_bio_8"</span>,  <span class="st">"wc2.1_2.5m_bio_9"</span>, <span class="st">"wc2.1_2.5m_bio_10"</span>, <span class="st">"wc2.1_2.5m_bio_11"</span>, <span class="st">"wc2.1_2.5m_bio_12"</span>, <span class="st">"wc2.1_2.5m_bio_13"</span>,<span class="st">"wc2.1_2.5m_bio_14"</span>, <span class="st">"wc2.1_2.5m_bio_15"</span>, <span class="st">"wc2.1_2.5m_bio_16"</span>, <span class="st">"wc2.1_2.5m_bio_17"</span>, <span class="st">"wc2.1_2.5m_bio_18"</span>, <span class="st">"wc2.1_2.5m_bio_19"</span> )</span>
<span id="cb25-234"><a href="#cb25-234" aria-hidden="true" tabindex="-1"></a>fut<span class="ot">&lt;-</span><span class="fu">crop</span>(fut, bioclimc)</span>
<span id="cb25-235"><a href="#cb25-235" aria-hidden="true" tabindex="-1"></a>max_predict_quly_2050<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_quly, <span class="at">data=</span>fut, <span class="at">filename =</span> <span class="st">"max_predict_quly_fut_2050_ssp585_ecearth3veg.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-236"><a href="#cb25-236" aria-hidden="true" tabindex="-1"></a>max_predict_qubi_2050<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_final_qubi, <span class="at">data=</span>fut, <span class="at">filename =</span> <span class="st">"max_predict_qubi_fut_2050_ssp585_ecearth3veg.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-237"><a href="#cb25-237" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_qubi_2050)</span>
<span id="cb25-238"><a href="#cb25-238" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_quly_current)</span>
<span id="cb25-239"><a href="#cb25-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-240"><a href="#cb25-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-241"><a href="#cb25-241" aria-hidden="true" tabindex="-1"></a><span class="co">#test with all bioclim variables qubi</span></span>
<span id="cb25-242"><a href="#cb25-242" aria-hidden="true" tabindex="-1"></a>data<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"quercusbicolor"</span>, bioclim_qubi, <span class="at">p=</span>qubi_pres, <span class="at">a=</span>qubi_bg )</span>
<span id="cb25-243"><a href="#cb25-243" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(qubi_train_all, qubi_test_all) <span class="sc">%&lt;-%</span> <span class="fu">trainValTest</span>(data, </span>
<span id="cb25-244"><a href="#cb25-244" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">test =</span> <span class="fl">0.2</span>, </span>
<span id="cb25-245"><a href="#cb25-245" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">only_presence =</span> <span class="cn">TRUE</span>, </span>
<span id="cb25-246"><a href="#cb25-246" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">seed =</span> <span class="dv">42</span>)</span>
<span id="cb25-247"><a href="#cb25-247" aria-hidden="true" tabindex="-1"></a><span class="co">#parition training data</span></span>
<span id="cb25-248"><a href="#cb25-248" aria-hidden="true" tabindex="-1"></a>cv<span class="ot">&lt;-</span><span class="fu">randomFolds</span>(qubi_train_all, <span class="at">k=</span><span class="dv">10</span>, <span class="at">only_presence=</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-249"><a href="#cb25-249" aria-hidden="true" tabindex="-1"></a><span class="co">#create maxent model with default values</span></span>
<span id="cb25-250"><a href="#cb25-250" aria-hidden="true" tabindex="-1"></a>maxent_allBC_qubi<span class="ot">&lt;-</span><span class="fu">train</span>(<span class="st">"Maxent"</span>, qubi_train_all, <span class="at">folds=</span>cv, <span class="at">progress=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-251"><a href="#cb25-251" aria-hidden="true" tabindex="-1"></a>max_allBC_qubi<span class="ot">&lt;-</span><span class="fu">combineCV</span>(maxent_allBC_qubi)</span>
<span id="cb25-252"><a href="#cb25-252" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training AUC: "</span>, <span class="fu">auc</span>(maxent_allBC_qubi))</span>
<span id="cb25-253"><a href="#cb25-253" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing AUC: "</span>, <span class="fu">auc</span>(maxent_allBC_qubi, <span class="at">test =</span> qubi_test_all))</span>
<span id="cb25-254"><a href="#cb25-254" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Training TSS: "</span>, <span class="fu">tss</span>(maxent_allBC_qubi))</span>
<span id="cb25-255"><a href="#cb25-255" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Testing TSS: "</span>, <span class="fu">tss</span>(maxent_allBC_qubi, <span class="at">test =</span> qubi_test_all))</span>
<span id="cb25-256"><a href="#cb25-256" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"D:/data/wc2.1_30s_bio/raw_rasters_touse/QUBI_and_QULY/modeling_outputs"</span>)</span>
<span id="cb25-257"><a href="#cb25-257" aria-hidden="true" tabindex="-1"></a>allBC_max_modelreport<span class="ot">&lt;-</span><span class="fu">modelReport</span>(max_allBC_qubi, <span class="at">folder=</span><span class="st">"max_allBC_qubi_modreport"</span>, <span class="at">test =</span> qubi_test_all, <span class="at">type =</span> <span class="st">"cloglog"</span>,<span class="at">response_curves =</span> <span class="cn">TRUE</span>,<span class="at">only_presence =</span> <span class="cn">TRUE</span>,<span class="at">jk =</span> <span class="cn">TRUE</span>,<span class="at">env =</span> bioclim_qubi, <span class="at">clamp =</span> <span class="cn">TRUE</span>, <span class="at">permut =</span> <span class="dv">10</span>,<span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-258"><a href="#cb25-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-259"><a href="#cb25-259" aria-hidden="true" tabindex="-1"></a><span class="co">#test with all bioclim variables quly</span></span>
<span id="cb25-260"><a href="#cb25-260" aria-hidden="true" tabindex="-1"></a>data<span class="ot">&lt;-</span><span class="fu">prepareSWD</span>(<span class="st">"quercuslyrata"</span>, bioclim_quly, <span class="at">p=</span>quly_pres, <span class="at">a=</span>quly_bg )</span>
<span id="cb25-261"><a href="#cb25-261" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(quly_train_all, quly_test_all) <span class="sc">%&lt;-%</span> <span class="fu">trainValTest</span>(data, </span>
<span id="cb25-262"><a href="#cb25-262" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">test =</span> <span class="fl">0.2</span>, </span>
<span id="cb25-263"><a href="#cb25-263" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">only_presence =</span> <span class="cn">TRUE</span>, </span>
<span id="cb25-264"><a href="#cb25-264" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">seed =</span> <span class="dv">42</span>)</span>
<span id="cb25-265"><a href="#cb25-265" aria-hidden="true" tabindex="-1"></a><span class="co">#parition training data</span></span>
<span id="cb25-266"><a href="#cb25-266" aria-hidden="true" tabindex="-1"></a>cv<span class="ot">&lt;-</span><span class="fu">randomFolds</span>(quly_train_all, <span class="at">k=</span><span class="dv">10</span>, <span class="at">only_presence=</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-267"><a href="#cb25-267" aria-hidden="true" tabindex="-1"></a><span class="co">#create maxent model with default values</span></span>
<span id="cb25-268"><a href="#cb25-268" aria-hidden="true" tabindex="-1"></a>maxent_allBC_quly<span class="ot">&lt;-</span><span class="fu">train</span>(<span class="st">"Maxent"</span>, quly_train_all, <span class="at">folds=</span>cv, <span class="at">progress=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-269"><a href="#cb25-269" aria-hidden="true" tabindex="-1"></a>max_allBC_quly<span class="ot">&lt;-</span><span class="fu">combineCV</span>(maxent_allBC_quly)</span>
<span id="cb25-270"><a href="#cb25-270" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"D:/data/wc2.1_30s_bio/raw_rasters_touse/QUBI_and_QULY/modeling_outputs"</span>)</span>
<span id="cb25-271"><a href="#cb25-271" aria-hidden="true" tabindex="-1"></a>allBC_quly_max_modelreport<span class="ot">&lt;-</span><span class="fu">modelReport</span>(max_allBC_quly, <span class="at">folder=</span><span class="st">"max_allBC_quly_modreport"</span>, <span class="at">test =</span> quly_test_all,<span class="at">type =</span> <span class="st">"cloglog"</span>,<span class="at">response_curves =</span> <span class="cn">TRUE</span>,<span class="at">only_presence =</span> <span class="cn">TRUE</span>,<span class="at">jk =</span> <span class="cn">TRUE</span>,<span class="at">env =</span> bioclim_quly, <span class="at">clamp =</span> <span class="cn">TRUE</span>, <span class="at">permut =</span> <span class="dv">10</span>,<span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-272"><a href="#cb25-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-273"><a href="#cb25-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-274"><a href="#cb25-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-275"><a href="#cb25-275" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_current<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_quly, <span class="at">data=</span>bioclimc, <span class="at">filename =</span> <span class="st">"max_allBC_predict_quly_current.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-276"><a href="#cb25-276" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_LGM<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_quly, <span class="at">data=</span>paleo, <span class="at">filename =</span> <span class="st">"max_allBC_predict_quly_LGM.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-277"><a href="#cb25-277" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_IG<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_quly, <span class="at">data=</span>interglacial, <span class="at">filename =</span> <span class="st">"max_allBC_predict_quly_IG.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-278"><a href="#cb25-278" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_midhol<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_quly, <span class="at">data=</span>midhol, <span class="at">filename =</span> <span class="st">"max_allBC_predict_quly_midholocene.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-279"><a href="#cb25-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-280"><a href="#cb25-280" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_current<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_qubi, <span class="at">data=</span>bioclimc, <span class="at">filename =</span> <span class="st">"max_allBC_predict_qubi_current.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-281"><a href="#cb25-281" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_LGM<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_qubi, <span class="at">data=</span>paleo, <span class="at">filename =</span> <span class="st">"max_allBC_predict_qubi_LGM.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-282"><a href="#cb25-282" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_IG<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_qubi, <span class="at">data=</span>interglacial, <span class="at">filename =</span> <span class="st">"max_allBC_predict_qubi_IG.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-283"><a href="#cb25-283" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_midhol<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_qubi, <span class="at">data=</span>midhol, <span class="at">filename =</span> <span class="st">"max_allBC_predict_qubi_midholocene.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-284"><a href="#cb25-284" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_quly_midhol)</span>
<span id="cb25-285"><a href="#cb25-285" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_qubi_midhol)</span>
<span id="cb25-286"><a href="#cb25-286" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_quly_LGM)</span>
<span id="cb25-287"><a href="#cb25-287" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_qubi_LGM)</span>
<span id="cb25-288"><a href="#cb25-288" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_quly_IG)</span>
<span id="cb25-289"><a href="#cb25-289" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_qubi_IG)</span>
<span id="cb25-290"><a href="#cb25-290" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_qubi_current)</span>
<span id="cb25-291"><a href="#cb25-291" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_quly_current)</span>
<span id="cb25-292"><a href="#cb25-292" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_2050<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_quly, <span class="at">data=</span>fut, <span class="at">filename =</span> <span class="st">"max_allBC_predict_quly_fut_2050_ssp585_ecearth3veg.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-293"><a href="#cb25-293" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_2050<span class="ot">&lt;-</span> SDMtune<span class="sc">::</span><span class="fu">predict</span>(<span class="at">object=</span>max_allBC_qubi, <span class="at">data=</span>fut, <span class="at">filename =</span> <span class="st">"max_allBC_predict_qubi_fut_2050_ssp585_ecearth3veg.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-294"><a href="#cb25-294" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_qubi_2050)</span>
<span id="cb25-295"><a href="#cb25-295" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(max_predict_allBC_quly_2050)</span>
<span id="cb25-296"><a href="#cb25-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-297"><a href="#cb25-297" aria-hidden="true" tabindex="-1"></a><span class="co">#get threhsolds (max training sens+spec)</span></span>
<span id="cb25-298"><a href="#cb25-298" aria-hidden="true" tabindex="-1"></a>ths_qubi<span class="ot">&lt;-</span>SDMtune<span class="sc">::</span><span class="fu">thresholds</span>(<span class="at">model=</span>max_allBC_qubi, <span class="at">test=</span>qubi_test_all)</span>
<span id="cb25-299"><a href="#cb25-299" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(ths_qubi,<span class="st">"D:/data/wc2.1_30s_bio/raw_rasters_touse/QUBI_and_QULY/modeling_outputs/thresholds_allBC_qubi.csv"</span>)</span>
<span id="cb25-300"><a href="#cb25-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-301"><a href="#cb25-301" aria-hidden="true" tabindex="-1"></a>ths_quly<span class="ot">&lt;-</span>SDMtune<span class="sc">::</span><span class="fu">thresholds</span>(<span class="at">model=</span>max_allBC_quly, <span class="at">test=</span>quly_test_all)</span>
<span id="cb25-302"><a href="#cb25-302" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(ths_quly,<span class="st">"D:/data/wc2.1_30s_bio/raw_rasters_touse/QUBI_and_QULY/modeling_outputs/thresholds_allBC_quly.csv"</span>)</span>
<span id="cb25-303"><a href="#cb25-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-304"><a href="#cb25-304" aria-hidden="true" tabindex="-1"></a>thr_qubi<span class="ot">&lt;-</span>max_allBC_qubi<span class="sc">@</span>model<span class="sc">@</span>results[<span class="st">'X10.percentile.training.presence.Cloglog.threshold'</span>,<span class="dv">1</span>]</span>
<span id="cb25-305"><a href="#cb25-305" aria-hidden="true" tabindex="-1"></a>thr_quly<span class="ot">&lt;-</span>max_allBC_quly<span class="sc">@</span>model<span class="sc">@</span>results[<span class="st">'X10.percentile.training.presence.Cloglog.threshold'</span>,<span class="dv">1</span>]</span>
<span id="cb25-306"><a href="#cb25-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-307"><a href="#cb25-307" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_current <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_qubi } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QUBI_current_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-308"><a href="#cb25-308" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_LGM <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_qubi } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QUBI_LGM_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-309"><a href="#cb25-309" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_IG <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_qubi } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QUBI_IG_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-310"><a href="#cb25-310" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_midhol <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_qubi } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QUBI_midhol_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-311"><a href="#cb25-311" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_qubi_2050 <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_qubi } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QUBI_2050_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-312"><a href="#cb25-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-313"><a href="#cb25-313" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_current <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_quly } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QULY_current_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-314"><a href="#cb25-314" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_LGM <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_quly } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QULY_LGM_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-315"><a href="#cb25-315" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_IG <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_quly } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QULY_IG_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-316"><a href="#cb25-316" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_midhol <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_quly } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QULY_midhol_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-317"><a href="#cb25-317" aria-hidden="true" tabindex="-1"></a>max_predict_allBC_quly_2050 <span class="sc">%&gt;%</span> { . <span class="sc">&gt;</span> thr_quly } <span class="sc">%&gt;%</span> <span class="fu">writeRaster</span>(., <span class="st">"QULY_2050_binary.tif"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-318"><a href="#cb25-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-319"><a href="#cb25-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-320"><a href="#cb25-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-321"><a href="#cb25-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-322"><a href="#cb25-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-323"><a href="#cb25-323" aria-hidden="true" tabindex="-1"></a><span class="co"># Define colors: # 0 (Neither) = White # 1 (QUBI only) = Blue # 2 (QULY only) = Green # 3 (Both) = Purple</span></span>
<span id="cb25-324"><a href="#cb25-324" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"D:/data/wc2.1_30s_bio/raw_rasters_touse/QUBI_and_QULY/modeling_outputs"</span>)</span>
<span id="cb25-325"><a href="#cb25-325" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"purple"</span>)</span>
<span id="cb25-326"><a href="#cb25-326" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(terra)</span>
<span id="cb25-327"><a href="#cb25-327" aria-hidden="true" tabindex="-1"></a>quly_fut<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QULY_2050_binary.tif"</span>)</span>
<span id="cb25-328"><a href="#cb25-328" aria-hidden="true" tabindex="-1"></a>qubi_fut<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QUBI_2050_binary.tif"</span>)</span>
<span id="cb25-329"><a href="#cb25-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-330"><a href="#cb25-330" aria-hidden="true" tabindex="-1"></a>quly_current<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QULY_current_binary.tif"</span>)</span>
<span id="cb25-331"><a href="#cb25-331" aria-hidden="true" tabindex="-1"></a>qubi_current<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QUBI_current_binary.tif"</span>)</span>
<span id="cb25-332"><a href="#cb25-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-333"><a href="#cb25-333" aria-hidden="true" tabindex="-1"></a>quly_midhol<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QULY_midhol_binary.tif"</span>)</span>
<span id="cb25-334"><a href="#cb25-334" aria-hidden="true" tabindex="-1"></a>qubi_midhol<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QUBI_midhol_binary.tif"</span>)</span>
<span id="cb25-335"><a href="#cb25-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-336"><a href="#cb25-336" aria-hidden="true" tabindex="-1"></a>quly_lgm<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QULY_LGM_binary.tif"</span>)</span>
<span id="cb25-337"><a href="#cb25-337" aria-hidden="true" tabindex="-1"></a>qubi_lgm<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QUBI_LGM_binary.tif"</span>)</span>
<span id="cb25-338"><a href="#cb25-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-339"><a href="#cb25-339" aria-hidden="true" tabindex="-1"></a>quly_ig<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QULY_IG_binary.tif"</span>)</span>
<span id="cb25-340"><a href="#cb25-340" aria-hidden="true" tabindex="-1"></a>qubi_ig<span class="ot">&lt;-</span><span class="fu">rast</span>(<span class="st">"QUBI_IG_binary.tif"</span>)</span>
<span id="cb25-341"><a href="#cb25-341" aria-hidden="true" tabindex="-1"></a>ig_overlap <span class="ot">&lt;-</span> qubi_ig <span class="sc">*</span> <span class="dv">1</span> <span class="sc">+</span> quly_ig <span class="sc">*</span> <span class="dv">2</span>  </span>
<span id="cb25-342"><a href="#cb25-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-343"><a href="#cb25-343" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the maps</span></span>
<span id="cb25-344"><a href="#cb25-344" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb25-345"><a href="#cb25-345" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(terra)</span>
<span id="cb25-346"><a href="#cb25-346" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb25-347"><a href="#cb25-347" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb25-348"><a href="#cb25-348" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geodata)</span>
<span id="cb25-349"><a href="#cb25-349" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(ig_overlap, <span class="at">xy=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-350"><a href="#cb25-350" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"presence"</span>)</span>
<span id="cb25-351"><a href="#cb25-351" aria-hidden="true" tabindex="-1"></a><span class="co"># Define color mapping</span></span>
<span id="cb25-352"><a href="#cb25-352" aria-hidden="true" tabindex="-1"></a>color_mapping <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"0"</span> <span class="ot">=</span> <span class="st">"white"</span>, <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"blue"</span>, <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"green"</span>, <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"purple"</span>)</span>
<span id="cb25-353"><a href="#cb25-353" aria-hidden="true" tabindex="-1"></a><span class="co"># Load US states boundaries from geodata</span></span>
<span id="cb25-354"><a href="#cb25-354" aria-hidden="true" tabindex="-1"></a>us_states <span class="ot">&lt;-</span> <span class="fu">gadm</span>(<span class="st">"USA"</span>, <span class="at">level=</span><span class="dv">1</span>, <span class="at">path=</span><span class="fu">tempdir</span>())  <span class="co"># Level 1 = states</span></span>
<span id="cb25-355"><a href="#cb25-355" aria-hidden="true" tabindex="-1"></a>us_sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(us_states)</span>
<span id="cb25-356"><a href="#cb25-356" aria-hidden="true" tabindex="-1"></a>can<span class="ot">&lt;-</span><span class="fu">gadm</span>(<span class="st">"Canada"</span>, <span class="at">level=</span><span class="dv">0</span>, <span class="at">path=</span><span class="fu">tempdir</span>())</span>
<span id="cb25-357"><a href="#cb25-357" aria-hidden="true" tabindex="-1"></a>can_sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(can)</span>
<span id="cb25-358"><a href="#cb25-358" aria-hidden="true" tabindex="-1"></a>raster_extent <span class="ot">&lt;-</span> <span class="fu">st_as_sfc</span>(<span class="fu">st_bbox</span>(<span class="fu">c</span>(<span class="at">xmin=</span><span class="fu">min</span>(df<span class="sc">$</span>x), <span class="at">xmax=</span><span class="fu">max</span>(df<span class="sc">$</span>x), <span class="at">ymin=</span><span class="fu">min</span>(df<span class="sc">$</span>y), <span class="at">ymax=</span><span class="fu">max</span>(df<span class="sc">$</span>y)), <span class="at">crs=</span><span class="fu">st_crs</span>(us_sf)))</span>
<span id="cb25-359"><a href="#cb25-359" aria-hidden="true" tabindex="-1"></a><span class="co"># Clip US states to raster extent</span></span>
<span id="cb25-360"><a href="#cb25-360" aria-hidden="true" tabindex="-1"></a>can_sf_clipped <span class="ot">&lt;-</span> <span class="fu">st_crop</span>(can_sf, raster_extent)</span>
<span id="cb25-361"><a href="#cb25-361" aria-hidden="true" tabindex="-1"></a>us_sf_clipped <span class="ot">&lt;-</span> <span class="fu">st_crop</span>(us_sf, raster_extent)</span>
<span id="cb25-362"><a href="#cb25-362" aria-hidden="true" tabindex="-1"></a>canusa<span class="ot">&lt;-</span><span class="fu">merge</span>(us_sf_clipped,can_sf_clipped)</span>
<span id="cb25-363"><a href="#cb25-363" aria-hidden="true" tabindex="-1"></a>single_sf <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(<span class="fu">list</span>(us_sf_clipped,can_sf_clipped))</span>
<span id="cb25-364"><a href="#cb25-364" aria-hidden="true" tabindex="-1"></a>dissolve_sf <span class="ot">&lt;-</span> <span class="fu">st_union</span>(single_sf)</span>
<span id="cb25-365"><a href="#cb25-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-366"><a href="#cb25-366" aria-hidden="true" tabindex="-1"></a>ig_overlap_map<span class="ot">&lt;-</span><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb25-367"><a href="#cb25-367" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">fill=</span><span class="fu">factor</span>(presence))) <span class="sc">+</span></span>
<span id="cb25-368"><a href="#cb25-368" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(</span>
<span id="cb25-369"><a href="#cb25-369" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span>color_mapping, </span>
<span id="cb25-370"><a href="#cb25-370" aria-hidden="true" tabindex="-1"></a>    <span class="at">name=</span><span class="st">"Species Presence"</span>,</span>
<span id="cb25-371"><a href="#cb25-371" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels=</span><span class="fu">c</span>(<span class="fu">expression</span>(<span class="st">"None"</span>), </span>
<span id="cb25-372"><a href="#cb25-372" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus bicolor"</span>)), </span>
<span id="cb25-373"><a href="#cb25-373" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus lyrata"</span>)), </span>
<span id="cb25-374"><a href="#cb25-374" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="st">"Hybrid Zone"</span>))</span>
<span id="cb25-375"><a href="#cb25-375" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-376"><a href="#cb25-376" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>us_sf_clipped, <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-377"><a href="#cb25-377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>x), <span class="fu">max</span>(df<span class="sc">$</span>x)), <span class="at">ylim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>y), <span class="fu">max</span>(df<span class="sc">$</span>y))) <span class="sc">+</span></span>
<span id="cb25-378"><a href="#cb25-378" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Interglacial (~130,000 years BP) Overlap Map"</span>, <span class="at">x=</span><span class="st">"Longitude"</span>, <span class="at">y=</span><span class="st">"Latitude"</span>) <span class="sc">+</span></span>
<span id="cb25-379"><a href="#cb25-379" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb25-380"><a href="#cb25-380" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb25-381"><a href="#cb25-381" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),             <span class="co"># Set all text to monospace</span></span>
<span id="cb25-382"><a href="#cb25-382" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),            <span class="co"># Monospace for title</span></span>
<span id="cb25-383"><a href="#cb25-383" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),       <span class="co"># Monospace for axis labels</span></span>
<span id="cb25-384"><a href="#cb25-384" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),        <span class="co"># Monospace for axis ticks</span></span>
<span id="cb25-385"><a href="#cb25-385" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>), </span>
<span id="cb25-386"><a href="#cb25-386" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>)      <span class="co"># Monospace for legend title</span></span>
<span id="cb25-387"><a href="#cb25-387" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-388"><a href="#cb25-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-389"><a href="#cb25-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-390"><a href="#cb25-390" aria-hidden="true" tabindex="-1"></a>lgm_overlap <span class="ot">&lt;-</span> qubi_lgm <span class="sc">*</span> <span class="dv">1</span> <span class="sc">+</span> quly_lgm <span class="sc">*</span> <span class="dv">2</span>  </span>
<span id="cb25-391"><a href="#cb25-391" aria-hidden="true" tabindex="-1"></a><span class="fu">freq</span>(lgm_overlap)</span>
<span id="cb25-392"><a href="#cb25-392" aria-hidden="true" tabindex="-1"></a>r_area <span class="ot">&lt;-</span> <span class="fu">cellSize</span>(lgm_overlap, <span class="at">unit =</span> <span class="st">"km"</span>)</span>
<span id="cb25-393"><a href="#cb25-393" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> lgm_overlap <span class="sc">==</span> <span class="dv">3</span></span>
<span id="cb25-394"><a href="#cb25-394" aria-hidden="true" tabindex="-1"></a>area_hyb <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-395"><a href="#cb25-395" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> lgm_overlap <span class="sc">!=</span> <span class="dv">0</span></span>
<span id="cb25-396"><a href="#cb25-396" aria-hidden="true" tabindex="-1"></a>area_tot <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-397"><a href="#cb25-397" aria-hidden="true" tabindex="-1"></a>lgm_df<span class="ot">&lt;-</span><span class="fu">c</span>(area_hyb, area_tot) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() </span>
<span id="cb25-398"><a href="#cb25-398" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(lgm_df)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"hybrid_area"</span>,<span class="st">"total_area_occupied"</span>)</span>
<span id="cb25-399"><a href="#cb25-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-400"><a href="#cb25-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-401"><a href="#cb25-401" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(area_3)</span>
<span id="cb25-402"><a href="#cb25-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-403"><a href="#cb25-403" aria-hidden="true" tabindex="-1"></a>lgm_overlap<span class="ot">&lt;-</span><span class="fu">crop</span>(lgm_overlap, dissolve_sf)</span>
<span id="cb25-404"><a href="#cb25-404" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(lgm_overlap, <span class="at">xy=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-405"><a href="#cb25-405" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"presence"</span>)</span>
<span id="cb25-406"><a href="#cb25-406" aria-hidden="true" tabindex="-1"></a><span class="co"># Define color mapping</span></span>
<span id="cb25-407"><a href="#cb25-407" aria-hidden="true" tabindex="-1"></a>lgm_overlap_map<span class="ot">&lt;-</span><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb25-408"><a href="#cb25-408" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">fill=</span><span class="fu">factor</span>(presence))) <span class="sc">+</span></span>
<span id="cb25-409"><a href="#cb25-409" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(</span>
<span id="cb25-410"><a href="#cb25-410" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span>color_mapping, </span>
<span id="cb25-411"><a href="#cb25-411" aria-hidden="true" tabindex="-1"></a>    <span class="at">name=</span><span class="st">"Species Presence"</span>,</span>
<span id="cb25-412"><a href="#cb25-412" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels=</span><span class="fu">c</span>(<span class="fu">expression</span>(<span class="st">"None"</span>), </span>
<span id="cb25-413"><a href="#cb25-413" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus bicolor"</span>)), </span>
<span id="cb25-414"><a href="#cb25-414" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus lyrata"</span>)), </span>
<span id="cb25-415"><a href="#cb25-415" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="st">"Hybrid Zone"</span>))</span>
<span id="cb25-416"><a href="#cb25-416" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-417"><a href="#cb25-417" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>us_sf_clipped, <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-418"><a href="#cb25-418" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>dissolve_sf, <span class="at">fill=</span><span class="st">'white'</span>, <span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-419"><a href="#cb25-419" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>x), <span class="fu">max</span>(df<span class="sc">$</span>x)), <span class="at">ylim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>y), <span class="fu">max</span>(df<span class="sc">$</span>y))) <span class="sc">+</span></span>
<span id="cb25-420"><a href="#cb25-420" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Last Glacial Maximum (~22,000 years BP) Climate"</span>, <span class="at">x=</span><span class="st">"Longitude"</span>, <span class="at">y=</span><span class="st">"Latitude"</span>) <span class="sc">+</span></span>
<span id="cb25-421"><a href="#cb25-421" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb25-422"><a href="#cb25-422" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb25-423"><a href="#cb25-423" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),             <span class="co"># Set all text to monospace</span></span>
<span id="cb25-424"><a href="#cb25-424" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),            <span class="co"># Monospace for title</span></span>
<span id="cb25-425"><a href="#cb25-425" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),       <span class="co"># Monospace for axis labels</span></span>
<span id="cb25-426"><a href="#cb25-426" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),        <span class="co"># Monospace for axis ticks</span></span>
<span id="cb25-427"><a href="#cb25-427" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>), </span>
<span id="cb25-428"><a href="#cb25-428" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>)      <span class="co"># Monospace for legend title</span></span>
<span id="cb25-429"><a href="#cb25-429" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-430"><a href="#cb25-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-431"><a href="#cb25-431" aria-hidden="true" tabindex="-1"></a>midhol_overlap <span class="ot">&lt;-</span> qubi_midhol <span class="sc">*</span> <span class="dv">1</span> <span class="sc">+</span> quly_midhol <span class="sc">*</span> <span class="dv">2</span>  </span>
<span id="cb25-432"><a href="#cb25-432" aria-hidden="true" tabindex="-1"></a><span class="fu">freq</span>(midhol_overlap)</span>
<span id="cb25-433"><a href="#cb25-433" aria-hidden="true" tabindex="-1"></a>r_area <span class="ot">&lt;-</span> <span class="fu">cellSize</span>(midhol_overlap, <span class="at">unit =</span> <span class="st">"km"</span>)</span>
<span id="cb25-434"><a href="#cb25-434" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> midhol_overlap <span class="sc">==</span> <span class="dv">3</span></span>
<span id="cb25-435"><a href="#cb25-435" aria-hidden="true" tabindex="-1"></a>area_hyb <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-436"><a href="#cb25-436" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> midhol_overlap <span class="sc">!=</span> <span class="dv">0</span></span>
<span id="cb25-437"><a href="#cb25-437" aria-hidden="true" tabindex="-1"></a>area_tot <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-438"><a href="#cb25-438" aria-hidden="true" tabindex="-1"></a>midhol_df<span class="ot">&lt;-</span><span class="fu">c</span>(area_hyb, area_tot) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() </span>
<span id="cb25-439"><a href="#cb25-439" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(midhol_df)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"hybrid_area"</span>,<span class="st">"total_area_occupied"</span>)</span>
<span id="cb25-440"><a href="#cb25-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-441"><a href="#cb25-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-442"><a href="#cb25-442" aria-hidden="true" tabindex="-1"></a>r_area <span class="ot">&lt;-</span> <span class="fu">cellSize</span>(current_overlap, <span class="at">unit =</span> <span class="st">"km"</span>)</span>
<span id="cb25-443"><a href="#cb25-443" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> current_overlap <span class="sc">==</span> <span class="dv">3</span></span>
<span id="cb25-444"><a href="#cb25-444" aria-hidden="true" tabindex="-1"></a>area_hyb <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-445"><a href="#cb25-445" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> current_overlap <span class="sc">!=</span> <span class="dv">0</span></span>
<span id="cb25-446"><a href="#cb25-446" aria-hidden="true" tabindex="-1"></a>area_tot <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-447"><a href="#cb25-447" aria-hidden="true" tabindex="-1"></a>current_df<span class="ot">&lt;-</span><span class="fu">c</span>(area_hyb, area_tot) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() </span>
<span id="cb25-448"><a href="#cb25-448" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(current_df)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"hybrid_area"</span>,<span class="st">"total_area_occupied"</span>)</span>
<span id="cb25-449"><a href="#cb25-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-450"><a href="#cb25-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-451"><a href="#cb25-451" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(midhol_overlap, <span class="at">xy=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-452"><a href="#cb25-452" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"presence"</span>)</span>
<span id="cb25-453"><a href="#cb25-453" aria-hidden="true" tabindex="-1"></a><span class="co"># Define color mapping</span></span>
<span id="cb25-454"><a href="#cb25-454" aria-hidden="true" tabindex="-1"></a>midhol_overlap_map<span class="ot">&lt;-</span><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb25-455"><a href="#cb25-455" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">fill=</span><span class="fu">factor</span>(presence))) <span class="sc">+</span></span>
<span id="cb25-456"><a href="#cb25-456" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(</span>
<span id="cb25-457"><a href="#cb25-457" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span>color_mapping, </span>
<span id="cb25-458"><a href="#cb25-458" aria-hidden="true" tabindex="-1"></a>    <span class="at">name=</span><span class="st">"Species Presence"</span>,</span>
<span id="cb25-459"><a href="#cb25-459" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels=</span><span class="fu">c</span>(<span class="fu">expression</span>(<span class="st">"None"</span>), </span>
<span id="cb25-460"><a href="#cb25-460" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus bicolor"</span>)), </span>
<span id="cb25-461"><a href="#cb25-461" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus lyrata"</span>)), </span>
<span id="cb25-462"><a href="#cb25-462" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="st">"Hybrid Zone"</span>))</span>
<span id="cb25-463"><a href="#cb25-463" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-464"><a href="#cb25-464" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>us_sf_clipped, <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-465"><a href="#cb25-465" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>x), <span class="fu">max</span>(df<span class="sc">$</span>x)), <span class="at">ylim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>y), <span class="fu">max</span>(df<span class="sc">$</span>y))) <span class="sc">+</span></span>
<span id="cb25-466"><a href="#cb25-466" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Mid-Holocene (~6,000 years BP) Climate"</span>, <span class="at">x=</span><span class="st">"Longitude"</span>, <span class="at">y=</span><span class="st">"Latitude"</span>) <span class="sc">+</span></span>
<span id="cb25-467"><a href="#cb25-467" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb25-468"><a href="#cb25-468" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb25-469"><a href="#cb25-469" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),             <span class="co"># Set all text to monospace</span></span>
<span id="cb25-470"><a href="#cb25-470" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),            <span class="co"># Monospace for title</span></span>
<span id="cb25-471"><a href="#cb25-471" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),       <span class="co"># Monospace for axis labels</span></span>
<span id="cb25-472"><a href="#cb25-472" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),        <span class="co"># Monospace for axis ticks</span></span>
<span id="cb25-473"><a href="#cb25-473" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>), </span>
<span id="cb25-474"><a href="#cb25-474" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>)      <span class="co"># Monospace for legend title</span></span>
<span id="cb25-475"><a href="#cb25-475" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-476"><a href="#cb25-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-477"><a href="#cb25-477" aria-hidden="true" tabindex="-1"></a>current_overlap <span class="ot">&lt;-</span> qubi_current <span class="sc">*</span> <span class="dv">1</span> <span class="sc">+</span> quly_current <span class="sc">*</span> <span class="dv">2</span>  </span>
<span id="cb25-478"><a href="#cb25-478" aria-hidden="true" tabindex="-1"></a><span class="fu">freq</span>(current_overlap)</span>
<span id="cb25-479"><a href="#cb25-479" aria-hidden="true" tabindex="-1"></a>r_area <span class="ot">&lt;-</span> <span class="fu">cellSize</span>(current_overlap, <span class="at">unit =</span> <span class="st">"km"</span>)</span>
<span id="cb25-480"><a href="#cb25-480" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> current_overlap <span class="sc">==</span> <span class="dv">3</span></span>
<span id="cb25-481"><a href="#cb25-481" aria-hidden="true" tabindex="-1"></a>area_hyb <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-482"><a href="#cb25-482" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> current_overlap <span class="sc">!=</span> <span class="dv">0</span></span>
<span id="cb25-483"><a href="#cb25-483" aria-hidden="true" tabindex="-1"></a>area_tot <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-484"><a href="#cb25-484" aria-hidden="true" tabindex="-1"></a>current_df<span class="ot">&lt;-</span><span class="fu">c</span>(area_hyb, area_tot) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() </span>
<span id="cb25-485"><a href="#cb25-485" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(current_df)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"hybrid_area"</span>,<span class="st">"total_area_occupied"</span>)</span>
<span id="cb25-486"><a href="#cb25-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-487"><a href="#cb25-487" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(current_overlap, <span class="at">xy=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-488"><a href="#cb25-488" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"presence"</span>)</span>
<span id="cb25-489"><a href="#cb25-489" aria-hidden="true" tabindex="-1"></a><span class="co"># Define color mapping</span></span>
<span id="cb25-490"><a href="#cb25-490" aria-hidden="true" tabindex="-1"></a>current_overlap_map<span class="ot">&lt;-</span><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb25-491"><a href="#cb25-491" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">fill=</span><span class="fu">factor</span>(presence))) <span class="sc">+</span></span>
<span id="cb25-492"><a href="#cb25-492" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(</span>
<span id="cb25-493"><a href="#cb25-493" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span>color_mapping, </span>
<span id="cb25-494"><a href="#cb25-494" aria-hidden="true" tabindex="-1"></a>    <span class="at">name=</span><span class="st">"Species Presence"</span>,</span>
<span id="cb25-495"><a href="#cb25-495" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels=</span><span class="fu">c</span>(<span class="fu">expression</span>(<span class="st">"None"</span>), </span>
<span id="cb25-496"><a href="#cb25-496" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus bicolor"</span>)), </span>
<span id="cb25-497"><a href="#cb25-497" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus lyrata"</span>)), </span>
<span id="cb25-498"><a href="#cb25-498" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="st">"Hybrid Zone"</span>))</span>
<span id="cb25-499"><a href="#cb25-499" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-500"><a href="#cb25-500" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>us_sf_clipped, <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-501"><a href="#cb25-501" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>x), <span class="fu">max</span>(df<span class="sc">$</span>x)), <span class="at">ylim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>y), <span class="fu">max</span>(df<span class="sc">$</span>y))) <span class="sc">+</span></span>
<span id="cb25-502"><a href="#cb25-502" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"1970-2000 Climate"</span>, <span class="at">x=</span><span class="st">"Longitude"</span>, <span class="at">y=</span><span class="st">"Latitude"</span>) <span class="sc">+</span></span>
<span id="cb25-503"><a href="#cb25-503" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb25-504"><a href="#cb25-504" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb25-505"><a href="#cb25-505" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),             <span class="co"># Set all text to monospace</span></span>
<span id="cb25-506"><a href="#cb25-506" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),            <span class="co"># Monospace for title</span></span>
<span id="cb25-507"><a href="#cb25-507" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),       <span class="co"># Monospace for axis labels</span></span>
<span id="cb25-508"><a href="#cb25-508" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),        <span class="co"># Monospace for axis ticks</span></span>
<span id="cb25-509"><a href="#cb25-509" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>), </span>
<span id="cb25-510"><a href="#cb25-510" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>)      <span class="co"># Monospace for legend title</span></span>
<span id="cb25-511"><a href="#cb25-511" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-512"><a href="#cb25-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-513"><a href="#cb25-513" aria-hidden="true" tabindex="-1"></a>fut_overlap <span class="ot">&lt;-</span> qubi_fut <span class="sc">*</span> <span class="dv">1</span> <span class="sc">+</span> quly_fut <span class="sc">*</span> <span class="dv">2</span>  </span>
<span id="cb25-514"><a href="#cb25-514" aria-hidden="true" tabindex="-1"></a>r_area <span class="ot">&lt;-</span> <span class="fu">cellSize</span>(fut_overlap, <span class="at">unit =</span> <span class="st">"km"</span>)</span>
<span id="cb25-515"><a href="#cb25-515" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> fut_overlap <span class="sc">==</span> <span class="dv">3</span></span>
<span id="cb25-516"><a href="#cb25-516" aria-hidden="true" tabindex="-1"></a>area_hyb <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-517"><a href="#cb25-517" aria-hidden="true" tabindex="-1"></a>mask_r <span class="ot">&lt;-</span> fut_overlap <span class="sc">!=</span> <span class="dv">0</span></span>
<span id="cb25-518"><a href="#cb25-518" aria-hidden="true" tabindex="-1"></a>area_tot <span class="ot">&lt;-</span> <span class="fu">global</span>(r_area <span class="sc">*</span> mask_r, <span class="at">fun =</span> <span class="st">"sum"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-519"><a href="#cb25-519" aria-hidden="true" tabindex="-1"></a>fut_df<span class="ot">&lt;-</span><span class="fu">c</span>(area_hyb, area_tot) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() </span>
<span id="cb25-520"><a href="#cb25-520" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(fut_df)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"hybrid_area"</span>,<span class="st">"total_area_occupied"</span>)</span>
<span id="cb25-521"><a href="#cb25-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-522"><a href="#cb25-522" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fut_overlap, <span class="at">xy=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-523"><a href="#cb25-523" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"presence"</span>)</span>
<span id="cb25-524"><a href="#cb25-524" aria-hidden="true" tabindex="-1"></a><span class="co"># Define color mapping</span></span>
<span id="cb25-525"><a href="#cb25-525" aria-hidden="true" tabindex="-1"></a>fut_overlap_map<span class="ot">&lt;-</span><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb25-526"><a href="#cb25-526" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">fill=</span><span class="fu">factor</span>(presence))) <span class="sc">+</span></span>
<span id="cb25-527"><a href="#cb25-527" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(</span>
<span id="cb25-528"><a href="#cb25-528" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span>color_mapping, </span>
<span id="cb25-529"><a href="#cb25-529" aria-hidden="true" tabindex="-1"></a>    <span class="at">name=</span><span class="st">"Species Presence"</span>,</span>
<span id="cb25-530"><a href="#cb25-530" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels=</span><span class="fu">c</span>(<span class="fu">expression</span>(<span class="st">"None"</span>), </span>
<span id="cb25-531"><a href="#cb25-531" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus bicolor"</span>)), </span>
<span id="cb25-532"><a href="#cb25-532" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"Quercus lyrata"</span>)), </span>
<span id="cb25-533"><a href="#cb25-533" aria-hidden="true" tabindex="-1"></a>             <span class="fu">expression</span>(<span class="st">"Hybrid Zone"</span>))</span>
<span id="cb25-534"><a href="#cb25-534" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-535"><a href="#cb25-535" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>us_sf_clipped, <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-536"><a href="#cb25-536" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>x), <span class="fu">max</span>(df<span class="sc">$</span>x)), <span class="at">ylim=</span><span class="fu">c</span>(<span class="fu">min</span>(df<span class="sc">$</span>y), <span class="fu">max</span>(df<span class="sc">$</span>y))) <span class="sc">+</span></span>
<span id="cb25-537"><a href="#cb25-537" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"2050 Climate RCP 8.5"</span>, <span class="at">x=</span><span class="st">"Longitude"</span>, <span class="at">y=</span><span class="st">"Latitude"</span>) <span class="sc">+</span></span>
<span id="cb25-538"><a href="#cb25-538" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb25-539"><a href="#cb25-539" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),<span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),<span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>),<span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">family=</span><span class="st">"mono"</span>)    </span>
<span id="cb25-540"><a href="#cb25-540" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb25-541"><a href="#cb25-541" aria-hidden="true" tabindex="-1"></a>names<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"lgm"</span>,<span class="st">"midhol"</span>,<span class="st">"current"</span>, <span class="st">"fut_df"</span>) </span>
<span id="cb25-542"><a href="#cb25-542" aria-hidden="true" tabindex="-1"></a>stats<span class="ot">&lt;-</span><span class="fu">rbind</span>(lgm_df,midhol_df,current_df,fut_df) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() </span>
<span id="cb25-543"><a href="#cb25-543" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(stats)<span class="ot">&lt;-</span>names</span>
<span id="cb25-544"><a href="#cb25-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-545"><a href="#cb25-545" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb25-546"><a href="#cb25-546" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb25-547"><a href="#cb25-547" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb25-548"><a href="#cb25-548" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grid)</span>
<span id="cb25-549"><a href="#cb25-549" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lattice)</span>
<span id="cb25-550"><a href="#cb25-550" aria-hidden="true" tabindex="-1"></a>plot_combined<span class="ot">&lt;-</span><span class="fu">arrangeGrob</span>(lgm_overlap_map, midhol_overlap_map, current_overlap_map, fut_overlap_map, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb25-551"><a href="#cb25-551" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as PNG</span></span>
<span id="cb25-552"><a href="#cb25-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-553"><a href="#cb25-553" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"hybridzone_maps.pdf"</span>, <span class="at">plot =</span> plot_combined, <span class="at">width =</span> <span class="dv">13</span>, <span class="at">height =</span> <span class="dv">13</span>, <span class="at">dpi =</span> <span class="dv">800</span>)</span>
<span id="cb25-554"><a href="#cb25-554" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-555"><a href="#cb25-555" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"C:/Users/jpark107/OneDrive - University of Tennessee/Desktop/Genetics_swo"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>